// @gooddata/goodstrap/lib/ColorPicker/ColorPicker.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty2 = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty2(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/define-property"));

var _defineProperties = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/define-properties"));

var _getOwnPropertyDescriptors = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptors"));

var _getOwnPropertyDescriptor = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptor"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/get-own-property-symbols"));

var _keys = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/keys"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty3 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _isEqual = _interopRequireDefault(require("lodash/isEqual"));

var _reactIntl = require("react-intl");

var _intlShape = require("../core/intlShape");

var _Button = _interopRequireDefault(require("../Button/Button"));

var _ColorPickerMatrix = _interopRequireDefault(require("./components/ColorPickerMatrix"));

var _HexColorInput = _interopRequireDefault(require("./components/HexColorInput"));

var _HueColorPicker = _interopRequireDefault(require("./components/HueColorPicker"));

var _ColorsPreview = _interopRequireDefault(require("./components/ColorsPreview"));

var _utils = require("./utils");

function ownKeys(object, enumerableOnly) { var keys = (0, _keys["default"])(object); if (_getOwnPropertySymbols["default"]) { var symbols = (0, _getOwnPropertySymbols["default"])(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return (0, _getOwnPropertyDescriptor["default"])(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty3["default"])(target, key, source[key]); }); } else if (_getOwnPropertyDescriptors["default"]) { (0, _defineProperties["default"])(target, (0, _getOwnPropertyDescriptors["default"])(source)); } else { ownKeys(Object(source)).forEach(function (key) { (0, _defineProperty2["default"])(target, key, (0, _getOwnPropertyDescriptor["default"])(source, key)); }); } } return target; }

var ColorPicker = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(ColorPicker, _Component);

  function ColorPicker(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, ColorPicker);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(ColorPicker).call(this, props));
    var currentHslColor = (0, _utils.getHslFromRgbColor)(_this.props.initialRgbColor);
    _this.state = {
      draftHslColor: currentHslColor
    };
    _this.onColorSelected = _this.onColorSelected.bind((0, _assertThisInitialized2["default"])(_this));
    _this.onHexInputColorSelected = _this.onHexInputColorSelected.bind((0, _assertThisInitialized2["default"])(_this));
    return _this;
  }

  (0, _createClass2["default"])(ColorPicker, [{
    key: "onColorSelected",
    value: function onColorSelected(selectedColor) {
      this.setState({
        draftHslColor: selectedColor
      });
    }
  }, {
    key: "onHexInputColorSelected",
    value: function onHexInputColorSelected(selectedColor) {
      if ((0, _utils.isHslColorBlackOrWhite)(selectedColor)) {
        this.setState({
          draftHslColor: _objectSpread({}, selectedColor, {
            h: this.state.draftHslColor.h
          })
        });
      } else {
        this.onColorSelected(selectedColor);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var currentHslColor = (0, _utils.getHslFromRgbColor)(this.props.initialRgbColor);
      var t = this.props.intl.formatMessage;
      return _react["default"].createElement("div", {
        className: "color-picker-container"
      }, _react["default"].createElement(_ColorPickerMatrix["default"], {
        initColor: this.state.draftHslColor,
        onColorSelected: this.onColorSelected
      }), _react["default"].createElement("div", {
        className: "color-picker-control-wrapper"
      }, _react["default"].createElement(_HueColorPicker["default"], {
        initColor: this.state.draftHslColor,
        onChange: this.onColorSelected
      }), _react["default"].createElement(_HexColorInput["default"], {
        initColor: this.state.draftHslColor,
        onInputChanged: this.onHexInputColorSelected,
        placeholder: t({
          id: "gs.color-picker.inputPlaceholder"
        }),
        label: t({
          id: "gs.color-picker.hex"
        })
      }), _react["default"].createElement(_ColorsPreview["default"], {
        currentHslColor: currentHslColor,
        draftHslColor: this.state.draftHslColor,
        currentTextLabel: t({
          id: "gs.color-picker.currentColor"
        }),
        draftTextLabel: t({
          id: "gs.color-picker.newColor"
        })
      }), _react["default"].createElement("div", {
        className: "color-picker-buttons-wrapper"
      }, _react["default"].createElement(_Button["default"], {
        value: t({
          id: "gs.color-picker.cancelButton"
        }),
        className: "gd-button-secondary gd-button color-picker-button",
        onClick: this.props.onCancel
      }), _react["default"].createElement(_Button["default"], {
        value: t({
          id: "gs.color-picker.okButton"
        }),
        disabled: (0, _isEqual["default"])(this.props.initialRgbColor, (0, _utils.getRgbFromHslColor)(this.state.draftHslColor)),
        className: "gd-button-action gd-button color-picker-ok-button",
        onClick: function onClick() {
          return _this2.props.onSubmit((0, _utils.getRgbFromHslColor)(_this2.state.draftHslColor));
        }
      }))));
    }
  }]);
  return ColorPicker;
}(_react.Component);

(0, _defineProperty3["default"])(ColorPicker, "propTypes", {
  initialRgbColor: _propTypes["default"].shape({
    r: _propTypes["default"].number,
    g: _propTypes["default"].number,
    b: _propTypes["default"].number
  }).isRequired,
  onSubmit: _propTypes["default"].func.isRequired,
  onCancel: _propTypes["default"].func.isRequired,
  intl: _intlShape.intlShape.isRequired
});

var _default = (0, _reactIntl.injectIntl)(ColorPicker);

exports["default"] = _default;
