// @gooddata/react-components/dist/components/visualizations/table/totals/AddTotalButton.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
// (C) 2007-2018 GoodData Corporation
var React = require("react");
var classNames = require("classnames");
var noop = require("lodash/noop");
var AddTotalButton = /** @class */ (function (_super) {
    __extends(AddTotalButton, _super);
    function AddTotalButton() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AddTotalButton.prototype.render = function () {
        var className = classNames({
            hidden: this.props.hidden,
        }, "s-total-add-row", "indigo-totals-add-row-button");
        return (React.createElement("div", { className: className, onClick: this.props.onClick, onMouseEnter: this.props.onMouseEnter, onMouseLeave: this.props.onMouseLeave }));
    };
    AddTotalButton.defaultProps = {
        onClick: noop,
        onMouseEnter: noop,
        onMouseLeave: noop,
        hidden: false,
    };
    return AddTotalButton;
}(React.PureComponent));
exports.AddTotalButton = AddTotalButton;
