// @gooddata/react-components/dist/internal/components/pluggableVisualizations/scatterPlot/PluggableScatterPlot.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
// (C) 2019-2020 GoodData Corporation
var React = require("react");
var react_dom_1 = require("react-dom");
var bucketConfig_1 = require("../../../utils/bucketConfig");
var cloneDeep = require("lodash/cloneDeep");
var PluggableBaseChart_1 = require("../baseChart/PluggableBaseChart");
var bucketHelper_1 = require("../../../utils/bucketHelper");
var BucketNames = require("../../../../constants/bucketNames");
var bucket_1 = require("../../../constants/bucket");
var sort_1 = require("../../../utils/sort");
var scatterPlotUiConfigHelper_1 = require("../../../utils/uiConfigHelpers/scatterPlotUiConfigHelper");
var uiConfig_1 = require("../../../constants/uiConfig");
var ScatterPlotConfigurationPanel_1 = require("../../configurationPanels/ScatterPlotConfigurationPanel");
var supportedProperties_1 = require("../../../constants/supportedProperties");
var propertiesHelper_1 = require("../../../utils/propertiesHelper");
var visualizationTypes_1 = require("../../../../constants/visualizationTypes");
var measureBucketItemsLimit = [
    {
        localIdentifier: BucketNames.MEASURES,
        itemsLimit: 1,
    },
    {
        localIdentifier: BucketNames.SECONDARY_MEASURES,
        itemsLimit: 1,
    },
];
exports.transformBuckets = function (buckets) {
    var bucketsWithLimitedMeasures = bucketHelper_1.limitNumberOfMeasuresInBuckets(buckets, 2, true);
    var measureBuckets = bucketHelper_1.transformMeasureBuckets(measureBucketItemsLimit, bucketsWithLimitedMeasures);
    var attributeBucket = {
        localIdentifier: BucketNames.ATTRIBUTE,
        items: bucketHelper_1.getAllAttributeItems(buckets).slice(0, 1),
    };
    return measureBuckets.concat([attributeBucket]);
};
var PluggableScatterPlot = /** @class */ (function (_super) {
    __extends(PluggableScatterPlot, _super);
    function PluggableScatterPlot(props) {
        var _this = _super.call(this, props) || this;
        _this.type = visualizationTypes_1.VisualizationTypes.SCATTER;
        _this.supportedPropertiesList = supportedProperties_1.SCATTERPLOT_SUPPORTED_PROPERTIES;
        _this.initializeProperties(props.visualizationProperties);
        return _this;
    }
    PluggableScatterPlot.prototype.getExtendedReferencePoint = function (referencePoint) {
        var clonedReferencePoint = cloneDeep(referencePoint);
        var newReferencePoint = __assign({}, clonedReferencePoint, { uiConfig: cloneDeep(uiConfig_1.DEFAULT_SCATTERPLOT_UICONFIG) });
        newReferencePoint = bucketHelper_1.removeAllArithmeticMeasuresFromDerived(newReferencePoint);
        newReferencePoint = bucketHelper_1.removeAllDerivedMeasures(newReferencePoint);
        newReferencePoint[bucket_1.BUCKETS] = exports.transformBuckets(newReferencePoint.buckets);
        newReferencePoint = scatterPlotUiConfigHelper_1.setScatterPlotUiConfig(newReferencePoint, this.intl, this.type);
        newReferencePoint = bucketConfig_1.configurePercent(newReferencePoint, true);
        newReferencePoint = bucketConfig_1.configureOverTimeComparison(newReferencePoint, !!this.featureFlags.enableWeekFilters);
        newReferencePoint = propertiesHelper_1.getReferencePointWithSupportedProperties(newReferencePoint, this.supportedPropertiesList);
        newReferencePoint = sort_1.removeSort(newReferencePoint);
        return Promise.resolve(bucketHelper_1.sanitizeFilters(newReferencePoint));
    };
    PluggableScatterPlot.prototype.renderConfigurationPanel = function () {
        if (document.querySelector(this.configPanelElement)) {
            react_dom_1.render(React.createElement(ScatterPlotConfigurationPanel_1.default, { locale: this.locale, references: this.references, properties: this.visualizationProperties, propertiesMeta: this.propertiesMeta, mdObject: this.mdObject, colors: this.colors, pushData: this.handlePushData, type: this.type, isError: this.isError, isLoading: this.isLoading, featureFlags: this.featureFlags }), document.querySelector(this.configPanelElement));
        }
    };
    return PluggableScatterPlot;
}(PluggableBaseChart_1.PluggableBaseChart));
exports.PluggableScatterPlot = PluggableScatterPlot;
