// @gooddata/goodstrap/lib/ColorPicker/components/ColorsPreview.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _utils = require("../utils");

// (C) 2007-2020 GoodData Corporation
var ColorsPreview = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2["default"])(ColorsPreview, _PureComponent);

  function ColorsPreview() {
    (0, _classCallCheck2["default"])(this, ColorsPreview);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(ColorsPreview).apply(this, arguments));
  }

  (0, _createClass2["default"])(ColorsPreview, [{
    key: "render",
    value: function render() {
      return _react["default"].createElement("div", {
        className: "color-picker-preview"
      }, _react["default"].createElement(ColorItem, {
        id: "current-color",
        hslColor: this.props.currentHslColor,
        textLabel: this.props.currentTextLabel
      }), _react["default"].createElement(ColorItem, {
        id: "new-color",
        hslColor: this.props.draftHslColor,
        textLabel: this.props.draftTextLabel
      }));
    }
  }]);
  return ColorsPreview;
}(_react.PureComponent);

exports["default"] = ColorsPreview;
(0, _defineProperty2["default"])(ColorsPreview, "propTypes", {
  currentHslColor: _propTypes["default"].shape({
    h: _propTypes["default"].number,
    s: _propTypes["default"].number,
    l: _propTypes["default"].number
  }).isRequired,
  draftHslColor: _propTypes["default"].shape({
    h: _propTypes["default"].number,
    s: _propTypes["default"].number,
    l: _propTypes["default"].number
  }).isRequired,
  currentTextLabel: _propTypes["default"].string,
  draftTextLabel: _propTypes["default"].string
});
(0, _defineProperty2["default"])(ColorsPreview, "defaultProps", {
  currentTextLabel: "current",
  draftTextLabel: "new"
});

var ColorItem = /*#__PURE__*/function (_PureComponent2) {
  (0, _inherits2["default"])(ColorItem, _PureComponent2);

  function ColorItem() {
    (0, _classCallCheck2["default"])(this, ColorItem);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(ColorItem).apply(this, arguments));
  }

  (0, _createClass2["default"])(ColorItem, [{
    key: "render",
    value: function render() {
      return _react["default"].createElement("div", {
        className: "color-picker-value-wrapper"
      }, _react["default"].createElement("div", {
        className: "color-value s-".concat(this.props.id),
        style: (0, _utils.getColorStyle)(this.props.hslColor)
      }), _react["default"].createElement("span", null, this.props.textLabel));
    }
  }]);
  return ColorItem;
}(_react.PureComponent);

(0, _defineProperty2["default"])(ColorItem, "propTypes", {
  id: _propTypes["default"].string.isRequired,
  hslColor: _propTypes["default"].shape({
    h: _propTypes["default"].number,
    s: _propTypes["default"].number,
    l: _propTypes["default"].number
  }).isRequired,
  textLabel: _propTypes["default"].string.isRequired
});
