// @gooddata/goodstrap/lib/ColorPicker/components/HexColorInput.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Input = _interopRequireDefault(require("../../Form/Input"));

var _utils = require("../utils");

// (C) 2007-2020 GoodData Corporation
var HexColorInput = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2["default"])(HexColorInput, _PureComponent);

  function HexColorInput() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2["default"])(this, HexColorInput);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2["default"])(this, (_getPrototypeOf2 = (0, _getPrototypeOf3["default"])(HexColorInput)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onInputChange", function (value) {
      if ((0, _utils.isHexColorValid)(value)) {
        var newHsl = (0, _utils.getHslFromHexColor)(value);

        _this.props.onInputChanged(newHsl);
      }
    });
    return _this;
  }

  (0, _createClass2["default"])(HexColorInput, [{
    key: "render",
    value: function render() {
      var hexValue = (0, _utils.getHexFromHslColor)(this.props.initColor);
      return _react["default"].createElement(_react.Fragment, null, _react["default"].createElement(_Input["default"], {
        className: "s-color-picker-hex",
        value: hexValue,
        onChange: this.onInputChange,
        placeholder: this.props.placeholder
      }), _react["default"].createElement("p", null, this.props.label));
    }
  }]);
  return HexColorInput;
}(_react.PureComponent);

exports["default"] = HexColorInput;
(0, _defineProperty2["default"])(HexColorInput, "propTypes", {
  initColor: _propTypes["default"].shape({
    h: _propTypes["default"].number,
    s: _propTypes["default"].number,
    l: _propTypes["default"].number
  }).isRequired,
  onInputChanged: _propTypes["default"].func.isRequired,
  placeholder: _propTypes["default"].string,
  label: _propTypes["default"].string
});
(0, _defineProperty2["default"])(HexColorInput, "defaultProps", {
  placeholder: "",
  label: ""
});
