// @gooddata/goodstrap/lib/ColorPicker/components/HueColorPicker.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _utils = require("../utils");

// (C) 2007-2020 GoodData Corporation

/* eslint-disable */

/**
 * Copyright (c) 2015 Case Sandberg
 * Copyrights licensed under the MIT License. See the accompanying LICENSE file for terms.
 */
var HueColorPicker = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2["default"])(HueColorPicker, _PureComponent);

  function HueColorPicker(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, HueColorPicker);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(HueColorPicker).call(this, props));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleChange", function (e) {
      var change = (0, _utils.calculateHueChange)(e, _this.props.initColor.h, _this.hueContainer.current);

      if (change && _this.props.onChange) {
        _this.props.onChange(change, e);
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleMouseDown", function (e) {
      _this.bindEventListeners();

      _this.handleChange(e);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleMouseUp", function () {
      _this.unbindEventListeners();
    });
    _this.hueContainer = _react["default"].createRef();
    return _this;
  }

  (0, _createClass2["default"])(HueColorPicker, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.unbindEventListeners();
    }
  }, {
    key: "getPointerStyle",
    value: function getPointerStyle() {
      return {
        position: "absolute",
        left: "".concat(this.props.initColor.h * 100 / 360, "%")
      };
    }
  }, {
    key: "bindEventListeners",
    value: function bindEventListeners() {
      window.addEventListener("mousemove", this.handleChange);
      window.addEventListener("mouseup", this.handleMouseUp);
    }
  }, {
    key: "unbindEventListeners",
    value: function unbindEventListeners() {
      window.removeEventListener("mousemove", this.handleChange);
      window.removeEventListener("mouseup", this.handleMouseUp);
    }
  }, {
    key: "render",
    value: function render() {
      return _react["default"].createElement("div", {
        className: "hue-picker hue-horizontal s-hue-picker",
        ref: this.hueContainer,
        onMouseDown: this.handleMouseDown,
        onMouseUp: this.handleMouseUp,
        onTouchMove: this.handleChange,
        onTouchStart: this.handleChange
      }, _react["default"].createElement("div", {
        style: this.getPointerStyle()
      }, _react["default"].createElement("div", {
        className: "color-picker-pointer s-color-picker-pointer"
      })));
    }
  }]);
  return HueColorPicker;
}(_react.PureComponent);

exports["default"] = HueColorPicker;
(0, _defineProperty2["default"])(HueColorPicker, "propTypes", {
  initColor: _propTypes["default"].shape({
    h: _propTypes["default"].number,
    s: _propTypes["default"].number,
    l: _propTypes["default"].number
  }).isRequired,
  onChange: _propTypes["default"].func.isRequired
});
