// @gooddata/react-components/dist/components/menu/Menu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// (C) 2007-2018 GoodData Corporation
var React = require("react");
var isFunction = require("lodash/isFunction");
var MenuState_1 = require("./MenuState");
var ControlledMenu_1 = require("./ControlledMenu");
require("../../../styles/css/menu.css");
var Menu = function (props) { return (React.createElement(MenuState_1.default, { opened: props.opened, defaultOpened: props.defaultOpened, onOpenedChange: props.onOpenedChange }, function (controlledProps) { return (React.createElement(ControlledMenu_1.default, { opened: controlledProps.opened, onOpenedChange: controlledProps.onOpenedChange, openAction: props.openAction, alignment: props.alignment, spacing: props.spacing, offset: props.offset, toggler: props.toggler, togglerWrapperClassName: props.togglerWrapperClassName, portalTarget: props.portalTarget, closeOnScroll: props.closeOnScroll }, isFunction(props.children)
    ? props.children({
        closeMenu: function () {
            return controlledProps.onOpenedChange({
                opened: false,
                source: "CLOSE_MENU_RENDER_PROP",
            });
        },
    })
    : props.children)); })); };
Menu.defaultProps = {
    openAction: "click",
    alignment: ["bottom", "right"],
    closeOnScroll: false,
};
exports.default = Menu;
