// @gooddata/goodstrap/lib/core/test/utils.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.renderIntoDocumentWithUnmount = renderIntoDocumentWithUnmount;
exports.withIntl = withIntl;
exports.intlMock = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _reactIntl = require("react-intl");

var _intl_formats = _interopRequireDefault(require("../../TestUtils/intl_formats"));

var _enUS = _interopRequireDefault(require("../../TestUtils/locales/en-US"));

// (C) 2007-2020 GoodData Corporation
function renderIntoDocumentWithUnmount(element) {
  var container = document.createElement("div"); // eslint-disable-next-line react/no-render-return-value

  var component = _reactDom["default"].render(element, container);

  component.unmount = function () {
    return _reactDom["default"].unmountComponentAtNode(container);
  };

  return component;
}

function withIntl(WrappedComponent) {
  var intlOptions = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {
    locale: "en",
    messages: _enUS["default"],
    formats: _intl_formats["default"]
  };
  return /*#__PURE__*/function (_Component) {
    (0, _inherits2["default"])(WithInt, _Component);

    function WithInt() {
      (0, _classCallCheck2["default"])(this, WithInt);
      return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(WithInt).apply(this, arguments));
    }

    (0, _createClass2["default"])(WithInt, [{
      key: "render",
      value: function render() {
        return _react["default"].createElement(_reactIntl.IntlProvider, intlOptions, _react["default"].createElement(WrappedComponent, this.props));
      }
    }]);
    return WithInt;
  }(_react.Component);
}

var intlMock = function intlMock() {
  var locale = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "en";
  return (0, _reactIntl.createIntl)({
    locale: locale,
    messages: _enUS["default"],
    formats: _intl_formats["default"]
  });
};

exports.intlMock = intlMock;
