// @gooddata/goodstrap/lib/@next/Dialog/ConfirmDialogBase.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Button = _interopRequireDefault(require("../../Button/Button"));

var _DialogBase2 = _interopRequireDefault(require("./DialogBase"));

// (C) 2020 GoodData Corporation
var ConfirmDialogBase = /*#__PURE__*/function (_DialogBase) {
  (0, _inherits2["default"])(ConfirmDialogBase, _DialogBase);

  function ConfirmDialogBase() {
    (0, _classCallCheck2["default"])(this, ConfirmDialogBase);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(ConfirmDialogBase).apply(this, arguments));
  }

  (0, _createClass2["default"])(ConfirmDialogBase, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          warning = _this$props.warning,
          isPositive = _this$props.isPositive,
          headline = _this$props.headline,
          displayCloseButton = _this$props.displayCloseButton,
          children = _this$props.children,
          onCancel = _this$props.onCancel,
          cancelButtonText = _this$props.cancelButtonText,
          onSubmit = _this$props.onSubmit,
          submitButtonText = _this$props.submitButtonText,
          isSubmitDisabled = _this$props.isSubmitDisabled;
      var dialogClasses = this.getDialogClasses("gd-confirm");
      var submitButtonClasses = (0, _classnames["default"])({
        "s-dialog-submit-button": true,
        "gd-button-action": isPositive,
        "gd-button-negative": !isPositive
      });
      return _react["default"].createElement("div", {
        tabIndex: 0,
        onKeyDown: this.onKeyDown
      }, _react["default"].createElement("div", {
        className: dialogClasses
      }, displayCloseButton && this.renderCloseButton(), _react["default"].createElement("div", {
        className: "gd-dialog-header"
      }, _react["default"].createElement("h3", null, headline)), !!warning && _react["default"].createElement("div", {
        className: "gd-dialog-warning"
      }, warning), _react["default"].createElement("div", {
        className: "gd-dialog-content"
      }, children), _react["default"].createElement("div", {
        className: "gd-dialog-footer"
      }, _react["default"].createElement(_Button["default"], {
        onClick: onCancel,
        className: "gd-button-secondary s-dialog-cancel-button",
        value: cancelButtonText
      }), _react["default"].createElement(_Button["default"], {
        onClick: onSubmit,
        className: submitButtonClasses,
        value: submitButtonText,
        disabled: isSubmitDisabled
      }))));
    }
  }]);
  return ConfirmDialogBase;
}(_DialogBase2["default"]);

exports["default"] = ConfirmDialogBase;
ConfirmDialogBase.defaultProps = {
  className: "",
  submitOnEnterKey: true,
  isSubmitDisabled: false,
  displayCloseButton: true,
  isPositive: true
};
