// @gooddata/goodstrap/lib/List/DateDatasetsListItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = exports.DateDatasetsListItem = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _defer = _interopRequireDefault(require("lodash/defer"));

var _reactIntl = require("react-intl");

var _classnames = _interopRequireDefault(require("classnames"));

var _jsUtils = require("@gooddata/js-utils");

var _ShortenedText = _interopRequireDefault(require("../core/ShortenedText"));

// (C) 2007-2020 GoodData Corporation
var DateDatasetsListItem = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(DateDatasetsListItem, _Component);

  function DateDatasetsListItem(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, DateDatasetsListItem);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(DateDatasetsListItem).call(this, props));
    _this.onDateDatasetChange = _this.onDateDatasetChange.bind((0, _assertThisInitialized2["default"])(_this));
    return _this;
  }

  (0, _createClass2["default"])(DateDatasetsListItem, [{
    key: "onDateDatasetChange",
    value: function onDateDatasetChange() {
      var _this$props = this.props,
          item = _this$props.item,
          onDateDatasetChange = _this$props.onDateDatasetChange; // Fix BB-1230 when overlay checks for clickOutside after react onChange
      // We want to defer onChage after overlay checked clickOutside

      (0, _defer["default"])(function () {
        return onDateDatasetChange(item);
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          item = _this$props2.item,
          activeDateDatasetId = _this$props2.activeDateDatasetId,
          unrelatedDateDatasetId = _this$props2.unrelatedDateDatasetId;

      if (item.type === "header") {
        return _react["default"].createElement("div", {
          className: "gd-list-item gd-list-item-header"
        }, _react["default"].createElement(_reactIntl.FormattedMessage, {
          id: item.title
        }));
      }

      var isSelected = activeDateDatasetId === item.id;
      var isUnrelated = unrelatedDateDatasetId === item.id;
      var classeNames = (0, _classnames["default"])("gd-list-item", "gd-list-item-shortened", "s-".concat(item.id), "s-".concat(_jsUtils.string.simplifyText(item.title)), {
        "is-selected": isSelected,
        "is-unrelated": isUnrelated
      });
      var tooltipAlignPoints = [{
        align: "cl cr",
        offset: {
          x: -10,
          y: 0
        }
      }, {
        align: "cr cl",
        offset: {
          x: 10,
          y: 0
        }
      }];
      return _react["default"].createElement("div", {
        className: classeNames,
        onClick: this.onDateDatasetChange
      }, _react["default"].createElement(_ShortenedText["default"], {
        tooltipAlignPoints: tooltipAlignPoints
      }, item.title));
    }
  }]);
  return DateDatasetsListItem;
}(_react.Component);

exports.DateDatasetsListItem = DateDatasetsListItem;
(0, _defineProperty2["default"])(DateDatasetsListItem, "propTypes", {
  activeDateDatasetId: _propTypes["default"].string,
  intl: _propTypes["default"].object.isRequired,
  item: _propTypes["default"].object.isRequired,
  onDateDatasetChange: _propTypes["default"].func.isRequired,
  unrelatedDateDatasetId: _propTypes["default"].string
});
(0, _defineProperty2["default"])(DateDatasetsListItem, "defaultProps", {
  activeDateDatasetId: "",
  unrelatedDateDatasetId: ""
});

var _default = (0, _reactIntl.injectIntl)(DateDatasetsListItem);

exports["default"] = _default;
