// @gooddata/goodstrap/lib/List/ListItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _get = _interopRequireDefault(require("lodash/get"));

// (C) 2007-2020 GoodData Corporation
var ListItem = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(ListItem, _Component);

  function ListItem() {
    (0, _classCallCheck2["default"])(this, ListItem);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(ListItem).apply(this, arguments));
  }

  (0, _createClass2["default"])(ListItem, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          item = _this$props.item,
          listItemClass = _this$props.listItemClass;
      var ListItemComponent = listItemClass;
      var itemType = (0, _get["default"])(item, "source.type", null);

      if (itemType === "separator") {
        return _react["default"].createElement("div", {
          className: "gd-list-item gd-list-item-separator"
        });
      }

      if (itemType === "header") {
        var itemTitle = (0, _get["default"])(item, "source.title", null);
        return _react["default"].createElement("div", {
          className: "gd-list-item gd-list-item-header"
        }, itemTitle);
      }

      return _react["default"].createElement(ListItemComponent, item);
    }
  }]);
  return ListItem;
}(_react.Component);

exports["default"] = ListItem;
(0, _defineProperty2["default"])(ListItem, "propTypes", {
  item: _propTypes["default"].object,
  // not required for lazy loaded lists
  listItemClass: _propTypes["default"].func.isRequired // TODO: Fix the naming here, class is not component

});
(0, _defineProperty2["default"])(ListItem, "defaultProps", {
  item: {}
});
