// @gooddata/goodstrap/lib/List/MenuList.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.Item = exports.Header = exports.Separator = exports.ItemsWrapper = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

// (C) 2007-2020 GoodData Corporation
var ItemsWrapper = function ItemsWrapper(_ref) {
  var smallItemsSpacing = _ref.smallItemsSpacing,
      children = _ref.children,
      restProps = (0, _objectWithoutProperties2["default"])(_ref, ["smallItemsSpacing", "children"]);
  return _react["default"].createElement("div", (0, _extends2["default"])({
    className: (0, _classnames["default"])({
      "gd-menu-wrapper": true,
      "gd-menu-wrapper-small-spacing": smallItemsSpacing
    })
  }, restProps), children);
};

exports.ItemsWrapper = ItemsWrapper;
ItemsWrapper.defaultProps = {
  smallItemsSpacing: false
};
ItemsWrapper.propTypes = {
  smallItemsSpacing: _propTypes["default"].bool,
  children: _propTypes["default"].node.isRequired
};

var Separator = function Separator(props) {
  return _react["default"].createElement("div", (0, _extends2["default"])({
    className: "gd-list-item gd-list-item-separator"
  }, props));
};

exports.Separator = Separator;

var Header = function Header(_ref2) {
  var children = _ref2.children,
      restProps = (0, _objectWithoutProperties2["default"])(_ref2, ["children"]);
  return _react["default"].createElement("div", (0, _extends2["default"])({
    className: "gd-list-item gd-list-item-header"
  }, restProps), children);
};

exports.Header = Header;
Header.propTypes = {
  children: _propTypes["default"].node.isRequired
};

var Item = function Item(_ref3) {
  var checked = _ref3.checked,
      subMenu = _ref3.subMenu,
      disabled = _ref3.disabled,
      children = _ref3.children,
      restProps = (0, _objectWithoutProperties2["default"])(_ref3, ["checked", "subMenu", "disabled", "children"]);
  return _react["default"].createElement("div", (0, _extends2["default"])({
    className: (0, _classnames["default"])({
      "gd-list-item": true,
      "gd-menu-item": true,
      "is-checked": checked,
      "is-submenu": subMenu,
      "is-disabled": disabled
    })
  }, restProps), children);
};

exports.Item = Item;
Item.defaultProps = {
  checked: false,
  subMenu: false,
  disabled: false
};
Item.propTypes = {
  checked: _propTypes["default"].bool,
  subMenu: _propTypes["default"].bool,
  disabled: _propTypes["default"].bool,
  children: _propTypes["default"].node.isRequired
};
