// @gooddata/goodstrap/lib/List/SearchableList.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/extends"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/assertThisInitialized"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _includes = _interopRequireDefault(require("lodash/includes"));

var _trim = _interopRequireDefault(require("lodash/trim"));

var _bindAll = _interopRequireDefault(require("lodash/bindAll"));

var _Input = _interopRequireDefault(require("../Form/Input"));

// (C) 2007-2020 GoodData Corporation
function getFilteredItems(items, searchString) {
  var preprocessedSearchString = (0, _trim["default"])(searchString).toLowerCase();
  return items.filter(function (item) {
    return (0, _includes["default"])(item.title.toLowerCase(), preprocessedSearchString);
  });
}

var EmptyResult = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(EmptyResult, _Component);

  function EmptyResult() {
    (0, _classCallCheck2["default"])(this, EmptyResult);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(EmptyResult).apply(this, arguments));
  }

  (0, _createClass2["default"])(EmptyResult, [{
    key: "render",
    value: function render() {
      return _react["default"].createElement("div", {
        className: "gd-list-noResults"
      }, this.props.text);
    }
  }]);
  return EmptyResult;
}(_react.Component);

EmptyResult.propTypes = {
  text: _propTypes["default"].string.isRequired
};

var SearchableList = /*#__PURE__*/function (_Component2) {
  (0, _inherits2["default"])(SearchableList, _Component2);

  function SearchableList(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, SearchableList);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(SearchableList).call(this, props));
    (0, _bindAll["default"])((0, _assertThisInitialized2["default"])(_this), "onSearchChange");
    _this.state = {
      itemsUnfiltered: props.items,
      itemsCount: props.itemsCount,
      items: props.items,
      searched: ""
    };
    return _this;
  }

  (0, _createClass2["default"])(SearchableList, [{
    key: "componentWillReceiveProps",
    value: function componentWillReceiveProps(nextProps) {
      this.setState({
        itemsUnfiltered: nextProps.items,
        items: this.applySearch(nextProps.items, this.state.searched)
      });
    }
  }, {
    key: "onSearchChange",
    value: function onSearchChange(searched) {
      var matchedItems = this.applySearch(this.state.itemsUnfiltered, searched);
      this.setState({
        items: matchedItems,
        itemsCount: matchedItems.length,
        searched: searched
      });
    }
  }, {
    key: "getClassNames",
    value: function getClassNames() {
      return (0, _classnames["default"])("searchable-list", this.props.className);
    }
  }, {
    key: "applySearch",
    value: function applySearch(items, searched) {
      return this.props.searchFn(items, searched);
    }
  }, {
    key: "renderList",
    value: function renderList() {
      var ListComponent = this.props.listClass;

      if (!this.state.items.length) {
        return _react["default"].createElement(EmptyResult, {
          text: this.props.noResultsText
        });
      }

      return _react["default"].createElement(ListComponent, (0, _extends2["default"])({}, this.props, {
        isFiltered: !!this.state.searched,
        itemsCount: this.state.itemsCount,
        items: this.state.items
      }));
    }
  }, {
    key: "render",
    value: function render() {
      // NOTE: may be split in the future
      return _react["default"].createElement("div", {
        className: this.getClassNames()
      }, _react["default"].createElement(_Input["default"], {
        autofocus: true,
        clearOnEsc: true,
        isSearch: true,
        isSmall: this.props.small,
        onChange: this.onSearchChange,
        value: this.state.searched,
        placeholder: this.props.placeholder
      }), this.renderList());
    }
  }]);
  return SearchableList;
}(_react.Component);

exports["default"] = SearchableList;
(0, _defineProperty2["default"])(SearchableList, "propTypes", {
  className: _propTypes["default"].string,
  items: _propTypes["default"].array.isRequired,
  itemsCount: _propTypes["default"].number.isRequired,
  height: _propTypes["default"].number.isRequired,
  itemHeight: _propTypes["default"].number.isRequired,
  listClass: _propTypes["default"].func.isRequired,
  placeholder: _propTypes["default"].string,
  noResultsText: _propTypes["default"].string,
  small: _propTypes["default"].bool,
  searchFn: _propTypes["default"].func
});
(0, _defineProperty2["default"])(SearchableList, "defaultProps", {
  className: "",
  noResultsText: "No matching data",
  placeholder: "",
  searchFn: getFilteredItems,
  small: false
});
