// @gooddata/goodstrap/lib/List/SingleSelectList.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _pick = _interopRequireDefault(require("lodash/pick"));

var _classnames = _interopRequireDefault(require("classnames"));

var _List = _interopRequireDefault(require("./List"));

var _ListItem = _interopRequireDefault(require("./ListItem"));

var _SingleSelectListItem = _interopRequireDefault(require("./SingleSelectListItem"));

var _Guid = require("../core/Guid");

// (C) 2007-2020 GoodData Corporation
var SingleSelectList = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(SingleSelectList, _Component);

  function SingleSelectList() {
    (0, _classCallCheck2["default"])(this, SingleSelectList);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(SingleSelectList).apply(this, arguments));
  }

  (0, _createClass2["default"])(SingleSelectList, [{
    key: "getSelectableItems",
    value: function getSelectableItems() {
      var props = this.props;
      return props.items.map(function (source) {
        return {
          source: source,
          onSelect: props.onSelect,
          onMouseOver: props.onItemMouseOver,
          onMouseOut: props.onItemMouseOut,
          onMouseEnter: props.onItemMouseEnter,
          onMouseLeave: props.onItemMouseLeave,
          selected: source === props.selection,
          id: props.getItemKey(source)
        };
      });
    }
  }, {
    key: "getClassNames",
    value: function getClassNames() {
      return (0, _classnames["default"])("gd-list", this.props.className);
    }
  }, {
    key: "getRowItem",
    value: function getRowItem() {
      var _this$props = this.props,
          rowItem = _this$props.rowItem,
          listItemClass = _this$props.listItemClass;
      return rowItem || _react["default"].createElement(_ListItem["default"], {
        listItemClass: listItemClass
      });
    }
  }, {
    key: "getDataSource",
    value: function getDataSource() {
      var selectableItems = this.getSelectableItems();
      return {
        rowsCount: this.props.itemsCount || selectableItems.length,
        getObjectAt: function getObjectAt(rowIndex) {
          return selectableItems[rowIndex];
        }
      };
    }
  }, {
    key: "render",
    value: function render() {
      var rowItem = this.getRowItem();
      var dataSource = this.getDataSource();
      return _react["default"].createElement(_List["default"], (0, _extends2["default"])({
        className: this.getClassNames()
      }, (0, _pick["default"])(this.props, ["width", "height", "itemHeight"]), {
        dataSource: dataSource,
        rowItem: rowItem,
        onScroll: this.props.onRangeChange,
        onScrollStart: this.props.onScrollStart,
        scrollToSelected: this.props.scrollToSelected
      }));
    }
  }]);
  return SingleSelectList;
}(_react.Component);

exports["default"] = SingleSelectList;
(0, _defineProperty2["default"])(SingleSelectList, "propTypes", {
  className: _propTypes["default"].string,
  getItemKey: _propTypes["default"].func,
  height: _propTypes["default"].number.isRequired,
  itemHeight: _propTypes["default"].number.isRequired,
  items: _propTypes["default"].array,
  itemsCount: _propTypes["default"].number,
  listItemClass: _propTypes["default"].oneOfType([_propTypes["default"].func, _propTypes["default"].instanceOf(_react.Component)]),
  onItemMouseOut: _propTypes["default"].func,
  onItemMouseOver: _propTypes["default"].func,
  onItemMouseEnter: _propTypes["default"].func,
  onItemMouseLeave: _propTypes["default"].func,
  onRangeChange: _propTypes["default"].func,
  onScrollStart: _propTypes["default"].func,
  onSelect: _propTypes["default"].func,
  scrollToSelected: _propTypes["default"].bool,
  rowItem: _propTypes["default"].element,
  selection: _propTypes["default"].object,
  width: _propTypes["default"].number.isRequired
});
(0, _defineProperty2["default"])(SingleSelectList, "defaultProps", {
  className: "",
  getItemKey: _Guid.guidFor,
  items: [],
  itemsCount: 0,
  listItemClass: _SingleSelectListItem["default"],
  // TODO add tests
  onItemMouseOut: function onItemMouseOut() {},
  onItemMouseOver: function onItemMouseOver() {},
  onItemMouseEnter: function onItemMouseEnter() {},
  onItemMouseLeave: function onItemMouseLeave() {},
  onRangeChange: function onRangeChange() {},
  onScrollStart: function onScrollStart() {},
  onSelect: function onSelect() {},
  rowItem: null,
  scrollToSelected: false,
  selection: {}
});
