// @gooddata/goodstrap/lib/core/FlexDimensions.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _throttle = _interopRequireDefault(require("lodash/throttle"));

var _pickBy = _interopRequireDefault(require("lodash/pickBy"));

var _domUtilities = require("./domUtilities");

// (C) 2007-2020 GoodData Corporation
var FlexDimensions = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(FlexDimensions, _Component);

  function FlexDimensions() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2["default"])(this, FlexDimensions);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2["default"])(this, (_getPrototypeOf2 = (0, _getPrototypeOf3["default"])(FlexDimensions)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "wrapperRef", (0, _react.createRef)());
    _this.state = {
      width: 0,
      height: 0
    };
    _this.throttledUpdateSize = (0, _throttle["default"])(_this.updateSize.bind((0, _assertThisInitialized2["default"])(_this)), 250, {
      leading: false
    });
    return _this;
  }

  (0, _createClass2["default"])(FlexDimensions, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      window.addEventListener("resize", this.throttledUpdateSize);
      this.throttledUpdateSize();
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.throttledUpdateSize.cancel();
      window.removeEventListener("resize", this.throttledUpdateSize);
    }
  }, {
    key: "getChildrenDimensions",
    value: function getChildrenDimensions() {
      var _this2 = this;

      return (0, _pickBy["default"])(this.state, function (_, key) {
        var setWidth = _this2.props.measureWidth && key === "width";
        var setHeight = _this2.props.measureHeight && key === "height";
        return setWidth || setHeight;
      });
    }
  }, {
    key: "updateSize",
    value: function updateSize() {
      var _elementRegion = (0, _domUtilities.elementRegion)(this.wrapperRef.current),
          width = _elementRegion.width,
          height = _elementRegion.height;

      this.setState({
        width: width,
        height: height
      });
    }
  }, {
    key: "renderChildren",
    value: function renderChildren() {
      var child = _react["default"].Children.only(this.props.children);

      return _react["default"].cloneElement(child, this.getChildrenDimensions());
    }
  }, {
    key: "render",
    value: function render() {
      var classNames = (0, _classnames["default"])(this.props.className);
      return _react["default"].createElement("div", {
        ref: this.wrapperRef,
        className: classNames
      }, this.renderChildren());
    }
  }]);
  return FlexDimensions;
}(_react.Component);

exports["default"] = FlexDimensions;
(0, _defineProperty2["default"])(FlexDimensions, "propTypes", {
  children: _propTypes["default"].node,
  className: _propTypes["default"].string,
  measureHeight: _propTypes["default"].bool,
  measureWidth: _propTypes["default"].bool
});
(0, _defineProperty2["default"])(FlexDimensions, "defaultProps", {
  children: false,
  className: "",
  measureWidth: true,
  measureHeight: true
});
