// @gooddata/goodstrap/lib/core/LoadingMask.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

// (C) 2007-2020 GoodData Corporation
var LoadingMask = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(LoadingMask, _Component);

  function LoadingMask() {
    (0, _classCallCheck2["default"])(this, LoadingMask);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(LoadingMask).apply(this, arguments));
  }

  (0, _createClass2["default"])(LoadingMask, [{
    key: "getClassNames",
    value: function getClassNames() {
      return (0, _classnames["default"])("s-isLoading", "loading-mask", this.props.classNames);
    }
  }, {
    key: "render",
    value: function render() {
      return _react["default"].createElement("div", {
        style: this.props.style,
        className: this.getClassNames()
      }, _react["default"].createElement("div", {
        className: "gd-spinner large"
      }));
    }
  }]);
  return LoadingMask;
}(_react.Component);

exports["default"] = LoadingMask;
(0, _defineProperty2["default"])(LoadingMask, "propTypes", {
  classNames: _propTypes["default"].string,
  style: _propTypes["default"].object
});
(0, _defineProperty2["default"])(LoadingMask, "defaultProps", {
  classNames: "",
  style: {}
});
