// @gooddata/goodstrap/lib/core/MediaQueries.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.IS_NOT_MOBILE_DEVICE = exports.IS_MOBILE_DEVICE = exports.IS_XXLARGE_ONLY = exports.IS_XXLARGE_UP = exports.IS_XLARGE_ONLY = exports.IS_XLARGE_UP = exports.IS_LARGE_ONLY = exports.IS_LARGE_UP = exports.IS_MEDIUM_ONLY = exports.IS_MEDIUM_UP = exports.IS_SMALL_ONLY = exports.IS_SMALL_UP = void 0;
// (C) 2007-2020 GoodData Corporation
var SCREEN = "only screen";
var ZERO_BREAKPOINT = 0;
var SMALL_BREAKPOINT = 640;
var MEDIUM_BREAKPOINT = 940;
var LARGE_BREAKPOINT = 1170;
var XLARGE_BREAKPOINT = 1590;
var XXLARGE_BREAKPOINT = 99999999;
var smallRange = {
  lower: ZERO_BREAKPOINT,
  upper: SMALL_BREAKPOINT
};
var mediumRange = {
  lower: SMALL_BREAKPOINT + 1,
  upper: MEDIUM_BREAKPOINT
};
var largeRange = {
  lower: MEDIUM_BREAKPOINT + 1,
  upper: LARGE_BREAKPOINT
};
var xlargeRange = {
  lower: LARGE_BREAKPOINT + 1,
  upper: XLARGE_BREAKPOINT
};
var xxlargeRange = {
  lower: XLARGE_BREAKPOINT + 1,
  upper: XXLARGE_BREAKPOINT
};

function isInRange(range) {
  return "".concat(SCREEN, " and (min-width:").concat(range.lower, "px) and (max-width:").concat(range.upper, "px)");
}

function isUpTo(range) {
  return "".concat(SCREEN, " and (min-width:").concat(range.lower, "px)");
}

var IS_SMALL_UP = isUpTo(smallRange);
exports.IS_SMALL_UP = IS_SMALL_UP;
var IS_SMALL_ONLY = isInRange(smallRange);
exports.IS_SMALL_ONLY = IS_SMALL_ONLY;
var IS_MEDIUM_UP = isUpTo(mediumRange);
exports.IS_MEDIUM_UP = IS_MEDIUM_UP;
var IS_MEDIUM_ONLY = isInRange(mediumRange);
exports.IS_MEDIUM_ONLY = IS_MEDIUM_ONLY;
var IS_LARGE_UP = isUpTo(largeRange);
exports.IS_LARGE_UP = IS_LARGE_UP;
var IS_LARGE_ONLY = isInRange(largeRange);
exports.IS_LARGE_ONLY = IS_LARGE_ONLY;
var IS_XLARGE_UP = isUpTo(xlargeRange);
exports.IS_XLARGE_UP = IS_XLARGE_UP;
var IS_XLARGE_ONLY = isInRange(xlargeRange);
exports.IS_XLARGE_ONLY = IS_XLARGE_ONLY;
var IS_XXLARGE_UP = isUpTo(xxlargeRange);
exports.IS_XXLARGE_UP = IS_XXLARGE_UP;
var IS_XXLARGE_ONLY = isInRange(xxlargeRange); // custom queries

exports.IS_XXLARGE_ONLY = IS_XXLARGE_ONLY;
var mobileRange = smallRange;
var notMobileRange = {
  lower: mediumRange.lower,
  upper: xxlargeRange.upper
};
var IS_MOBILE_DEVICE = isInRange(mobileRange);
exports.IS_MOBILE_DEVICE = IS_MOBILE_DEVICE;
var IS_NOT_MOBILE_DEVICE = isInRange(notMobileRange);
exports.IS_NOT_MOBILE_DEVICE = IS_NOT_MOBILE_DEVICE;
