// @gooddata/goodstrap/lib/core/Tabs.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactIntl = require("react-intl");

var _classnames = _interopRequireDefault(require("classnames"));

var _get = _interopRequireDefault(require("lodash/get"));

var _jsUtils = require("@gooddata/js-utils");

// (C) 2007-2020 GoodData Corporation
var Tabs = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(Tabs, _Component);

  function Tabs(props) {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2["default"])(this, Tabs);

    for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      args[_key - 1] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2["default"])(this, (_getPrototypeOf2 = (0, _getPrototypeOf3["default"])(Tabs)).call.apply(_getPrototypeOf2, [this, props].concat(args)));
    _this.state = {
      selectedTabId: props.selectedTabId || (0, _get["default"])(props, ["tabs", 0, "id"])
    };
    return _this;
  }

  (0, _createClass2["default"])(Tabs, [{
    key: "selectTab",
    value: function selectTab(tab) {
      var selectedTabId = (0, _get["default"])(tab, "id");
      var noChange = selectedTabId === this.state.selectedTabId;

      if (noChange) {
        return;
      }

      this.setState({
        selectedTabId: selectedTabId
      });
      this.props.onTabSelect(tab);
    }
  }, {
    key: "renderTab",
    value: function renderTab(tab) {
      var _this2 = this;

      var tabClassName = (0, _classnames["default"])((0, _defineProperty2["default"])({
        "is-active": tab.id === this.state.selectedTabId,
        "gd-tab": true
      }, "s-".concat(_jsUtils.string.simplifyText(tab.id)), true));
      return _react["default"].createElement("div", {
        className: tabClassName,
        key: tab.id,
        onClick: function onClick() {
          _this2.selectTab(tab);
        }
      }, _react["default"].createElement("span", null, _react["default"].createElement(_reactIntl.FormattedMessage, {
        id: tab.id
      })));
    }
  }, {
    key: "renderTabs",
    value: function renderTabs() {
      return this.props.tabs.map(this.renderTab, this);
    }
  }, {
    key: "render",
    value: function render() {
      var classNames = (0, _classnames["default"])(this.props.className, {
        "gd-tabs": true,
        small: true
      });
      return _react["default"].createElement("div", {
        className: classNames
      }, this.renderTabs());
    }
  }]);
  return Tabs;
}(_react.Component);

exports["default"] = Tabs;
(0, _defineProperty2["default"])(Tabs, "propTypes", {
  className: _propTypes["default"].string,
  onTabSelect: _propTypes["default"].func,
  selectedTabId: _propTypes["default"].string,
  tabs: _propTypes["default"].arrayOf(_propTypes["default"].shape({
    id: _propTypes["default"].string
  }))
});
(0, _defineProperty2["default"])(Tabs, "defaultProps", {
  className: "",
  onTabSelect: function onTabSelect() {},
  selectedTabId: "",
  tabs: []
});
