// @gooddata/goodstrap/lib/@next/Header/colors.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.toColor = toColor;
exports.luminanceOf = luminanceOf;
exports.chooseColor = chooseColor;
exports.getTextColor = getTextColor;
exports.getItemActiveColor = getItemActiveColor;
exports.getWorkspacePickerHoverColor = getWorkspacePickerHoverColor;
exports.getItemHoverColor = getItemHoverColor;
exports.getSeparatorColor = getSeparatorColor;

var _parseInt2 = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/parse-int"));

// (C) 2007-2020 GoodData Corporation
var HEX3_REGEX = /^#([\da-f])([\da-f])([\da-f])/i;
var HEX6_REGEX = /^#([\da-fA-F]{2})([\da-fA-F]{2})([\da-fA-F]{2})/;
/**
 * Helper function that transforms any string representation of a color
 * into dojo.Color object.
 *
 * @param String string
 * @return dojo.Color
 */

function toColor(hexColor) {
  var parsed;
  var p = _parseInt2["default"];
  var color = hexColor.replace(/\s*/g, ""); // Remove all spaces

  var hex3Match = HEX3_REGEX.exec(color);
  var hex6Match = HEX6_REGEX.exec(color);

  if (hex6Match) {
    // 6 digit hex
    parsed = [p(hex6Match[1], 16), p(hex6Match[2], 16), p(hex6Match[3], 16)];
  } else if (hex3Match) {
    // 3 digit hex
    parsed = [p(hex3Match[1], 16) * 17, p(hex3Match[2], 16) * 17, p(hex3Match[3], 16) * 17];
  }

  return {
    r: parsed[0],
    g: parsed[1],
    b: parsed[2],
    a: 1
  };
}
/**
 * Helper function that returns a luminance of a color.
 * Color must be an object with r, g and b properties.
 *
 * @param  {dojo.Color} color
 * @return {Number} Luminance in interval 0 - 255
 */


function luminanceOf(color) {
  // Imprecise luminance values may arise from finite precision of RGB coefitients
  // We have to round this value to two decimal places to avoid strange results
  var raw = 0.2126 * color.r + 0.7152 * color.g + 0.0722 * color.b;
  var rounded = Math.round(raw * 100) / 100;
  return rounded;
}
/**
 * Function that chooses from two colors based on luminance
 * of control color. This luminance is compared to specified threshold value.
 *
 * @param  {String} color as a String
 * @param  {Number} threshold Luminance threshold value in interval 1 - 255
 * @param  {String} dark Dark color variant for when control color luminance is higher than threshold
 * @param  {String} light Light color variant for when control color luminance is lower or equal than threshold
 * @return {String}
 */


function chooseColor(hexColor, threshold, dark, light) {
  if (!hexColor) {
    return "";
  }

  return luminanceOf(toColor(hexColor)) > threshold ? dark : light;
}

function getTextColor(headerTextColor, headerColor) {
  var detectedColor = chooseColor(headerColor, 128, "#000", "#FFF");
  return headerTextColor || detectedColor;
}

function getItemActiveColor(activeColor, headerColor) {
  var detectedColor = chooseColor(headerColor, 128, "#000", "#FFF");
  return activeColor || detectedColor;
}

function getWorkspacePickerHoverColor(headerColor) {
  var detectedColor = chooseColor(headerColor, 254.99, "rgba(0, 0, 0, .02)", "");
  return detectedColor || "";
}

function getItemHoverColor(headerColor, activeColor) {
  var lightOpacity = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0.5;
  var darkOpacity = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 0.25;
  var backgroundColorLuminance = luminanceOf(toColor(headerColor || "#000"));
  var emphasisColorLuminance = luminanceOf(toColor(activeColor || "#fff"));
  return emphasisColorLuminance > backgroundColorLuminance ? "rgba(255, 255, 255, ".concat(lightOpacity, ")") : "rgba(0, 0, 0, ".concat(darkOpacity, ")");
}

function getSeparatorColor(headerColor, activeColor) {
  return getItemHoverColor(headerColor, activeColor, 0.25, 0.08);
}
