// @gooddata/goodstrap/lib/@next/Header/HeaderWorkspacePickerItem.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.HeaderWorkspacePickerItem = exports.CoreHeaderWorkspacePickerItem = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _reactIntl = require("react-intl");

var _classnames = _interopRequireDefault(require("classnames"));

var _jsUtils = require("@gooddata/js-utils");

// (C) 2007-2020 GoodData Corporation
var CoreHeaderWorkspacePickerItem = function CoreHeaderWorkspacePickerItem(_ref) {
  var _cx;

  var intl = _ref.intl,
      title = _ref.title,
      isLoading = _ref.isLoading,
      isSelected = _ref.isSelected,
      isDemo = _ref.isDemo,
      onClick = _ref.onClick;
  var t = intl.formatMessage;

  if (isLoading) {
    return _react["default"].createElement("div", {
      className: "gd-list-item gd-project-list-item gd-is-loading"
    }, _react["default"].createElement("div", {
      className: "gd-spinner small"
    }));
  }

  var classes = (0, _classnames["default"])((_cx = {
    "gd-list-item": true,
    "gd-project-list-item": true
  }, (0, _defineProperty2["default"])(_cx, "s-".concat(_jsUtils.string.simplifyText(title)), true), (0, _defineProperty2["default"])(_cx, "gd-is-selected", isSelected), (0, _defineProperty2["default"])(_cx, "gd-is-demo", isDemo), _cx));
  return _react["default"].createElement("div", {
    className: classes,
    title: title,
    onClick: onClick
  }, _react["default"].createElement("span", {
    className: "project-title"
  }, title), isDemo && _react["default"].createElement("span", {
    className: "demo-sticker"
  }, t({
    id: "gs.header.projectPicker.demo"
  })));
};

exports.CoreHeaderWorkspacePickerItem = CoreHeaderWorkspacePickerItem;
var HeaderWorkspacePickerItem = (0, _reactIntl.injectIntl)(CoreHeaderWorkspacePickerItem);
exports.HeaderWorkspacePickerItem = HeaderWorkspacePickerItem;
