// @gooddata/react-components/dist/components/filters/DateFilter/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// (C) 2007-2019 GoodData Corporation
var i18n_1 = require("./constants/i18n");
var config_1 = require("./constants/config");
var OptionValidation_1 = require("./validation/OptionValidation");
var AFMConversions_1 = require("./utils/AFMConversions");
var PeriodExlusion_1 = require("./utils/PeriodExlusion");
var DateFilterTitle_1 = require("./utils/Translations/DateFilterTitle");
var DateFilter_1 = require("./DateFilter");
exports.DateFilter = DateFilter_1.DateFilter;
var DateFilterHelpers = {
    validateFilterOption: OptionValidation_1.validateFilterOption,
    getDateFilterTitle: DateFilterTitle_1.getDateFilterTitle,
    getDateFilterTitleUsingTranslator: DateFilterTitle_1.getDateFilterTitleUsingTranslator,
    getDateFilterRepresentation: DateFilterTitle_1.getDateFilterRepresentation,
    granularityIntlCodes: i18n_1.granularityIntlCodes,
    applyExcludeCurrentPeriod: PeriodExlusion_1.applyExcludeCurrentPeriod,
    defaultDateFilterOptions: config_1.defaultDateFilterOptions,
    canExcludeCurrentPeriod: PeriodExlusion_1.canExcludeCurrentPeriod,
    mapOptionToAfm: AFMConversions_1.mapOptionToAfm,
};
exports.DateFilterHelpers = DateFilterHelpers;
