// @gooddata/goodstrap/lib/DocumentHeader/DocumentHeader.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactDocumentTitle = _interopRequireDefault(require("react-document-title"));

var _AppleTouchIcon = _interopRequireDefault(require("./AppleTouchIcon"));

var _Favicon = _interopRequireDefault(require("./Favicon"));

// (C) 2007-2020 GoodData Corporation
var DocumentHeader = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(DocumentHeader, _Component);

  function DocumentHeader() {
    (0, _classCallCheck2["default"])(this, DocumentHeader);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(DocumentHeader).apply(this, arguments));
  }

  (0, _createClass2["default"])(DocumentHeader, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          pageTitle = _this$props.pageTitle,
          brandTitle = _this$props.brandTitle,
          appleTouchIconUrl = _this$props.appleTouchIconUrl,
          faviconUrl = _this$props.faviconUrl;
      var title = brandTitle ? "".concat(pageTitle, " - ").concat(brandTitle) : pageTitle;
      return _react["default"].createElement("div", null, _react["default"].createElement(_reactDocumentTitle["default"], {
        title: title
      }), _react["default"].createElement(_AppleTouchIcon["default"], {
        url: appleTouchIconUrl
      }), _react["default"].createElement(_Favicon["default"], {
        url: faviconUrl
      }));
    }
  }]);
  return DocumentHeader;
}(_react.Component);

exports["default"] = DocumentHeader;
(0, _defineProperty2["default"])(DocumentHeader, "propTypes", {
  pageTitle: _propTypes["default"].string.isRequired,
  brandTitle: _propTypes["default"].string,
  appleTouchIconUrl: _propTypes["default"].string,
  faviconUrl: _propTypes["default"].string
});
(0, _defineProperty2["default"])(DocumentHeader, "defaultProps", {
  brandTitle: "",
  appleTouchIconUrl: "",
  faviconUrl: ""
});
