// @gooddata/goodstrap/lib/@next/AutoSize/AutoSize.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.AutoSize = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _throttle = _interopRequireDefault(require("lodash/throttle"));

var _domUtilities = require("../../core/domUtilities");

// (C) 2007-2020 GoodData Corporation
// tslint:disable member-ordering
var AutoSize = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(AutoSize, _Component);

  function AutoSize() {
    var _this;

    (0, _classCallCheck2["default"])(this, AutoSize);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(AutoSize).apply(this, arguments));
    _this.state = {
      width: 0,
      height: 0
    };

    _this.updateSize = function () {
      var _elementRegion = (0, _domUtilities.elementRegion)(_this.wrapperRef.current),
          width = _elementRegion.width,
          height = _elementRegion.height;

      _this.setState({
        width: width,
        height: height
      });
    };

    _this.throttledUpdateSize = (0, _throttle["default"])(_this.updateSize, 250, {
      leading: false
    });
    _this.wrapperRef = (0, _react.createRef)();
    return _this;
  }

  (0, _createClass2["default"])(AutoSize, [{
    key: "render",
    value: function render() {
      var children = this.props.children;
      var _this$state = this.state,
          width = _this$state.width,
          height = _this$state.height;
      return _react["default"].createElement("div", {
        ref: this.wrapperRef,
        style: {
          height: "100%",
          width: "100%"
        }
      }, children({
        width: width,
        height: height
      }));
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      window.addEventListener("resize", this.throttledUpdateSize);
      this.throttledUpdateSize();
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.throttledUpdateSize.cancel();
      window.removeEventListener("resize", this.throttledUpdateSize);
    }
  }]);
  return AutoSize;
}(_react.Component);

exports.AutoSize = AutoSize;
