// @gooddata/react-components/dist/components/filters/DateFilter/utils/AFMConversions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// (C) 2007-2019 GoodData Corporation
var typings_1 = require("@gooddata/typings");
var PeriodExlusion_1 = require("./PeriodExlusion");
exports.mapAbsoluteFilterToAfm = function (value, dataSet) { return ({
    absoluteDateFilter: {
        dataSet: dataSet,
        from: value.from,
        to: value.to,
    },
}); };
exports.mapRelativeFilterToAfm = function (value, dataSet) { return ({
    relativeDateFilter: {
        dataSet: dataSet,
        from: value.from,
        to: value.to,
        granularity: value.granularity,
    },
}); };
exports.mapOptionToAfm = function (value, dateDataSet, excludeCurrentPeriod) {
    var excludeApplied = PeriodExlusion_1.applyExcludeCurrentPeriod(value, excludeCurrentPeriod);
    if (typings_1.ExtendedDateFilters.isAllTimeDateFilter(excludeApplied)) {
        return null;
    }
    if (typings_1.ExtendedDateFilters.isAbsoluteDateFilterOption(excludeApplied)) {
        return exports.mapAbsoluteFilterToAfm(excludeApplied, dateDataSet);
    }
    if (typings_1.ExtendedDateFilters.isRelativeDateFilterOption(excludeApplied)) {
        return exports.mapRelativeFilterToAfm(excludeApplied, dateDataSet);
    }
    throw new Error("Uknown date filter value provided");
};
