// @gooddata/goodstrap/lib/Dropdown/DropdownButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Button = _interopRequireDefault(require("../Button/Button"));

// (C) 2007-2020 GoodData Corporation
// eslint-disable-next-line import/no-unresolved, import/extensions
var DropdownButton = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(DropdownButton, _Component);

  function DropdownButton() {
    (0, _classCallCheck2["default"])(this, DropdownButton);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(DropdownButton).apply(this, arguments));
  }

  (0, _createClass2["default"])(DropdownButton, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          isOpen = _this$props.isOpen,
          isSmall = _this$props.isSmall,
          title = _this$props.title,
          value = _this$props.value,
          disabled = _this$props.disabled,
          className = _this$props.className,
          iconLeft = _this$props.iconLeft;
      var buttonClasses = (0, _classnames["default"])({
        "gd-button-primary": true,
        "gd-button-small": isSmall,
        "button-dropdown": true,
        "is-dropdown-open": isOpen,
        "is-active": isOpen,
        "dropdown-button": true
      }, className);
      return _react["default"].createElement(_Button["default"], {
        title: title,
        className: buttonClasses,
        value: value,
        iconLeft: iconLeft,
        iconRight: isOpen ? "icon-navigateup" : "icon-navigatedown",
        disabled: disabled
      });
    }
  }]);
  return DropdownButton;
}(_react.Component);

exports["default"] = DropdownButton;
(0, _defineProperty2["default"])(DropdownButton, "propTypes", {
  className: _propTypes["default"].string,
  disabled: _propTypes["default"].bool,
  isOpen: _propTypes["default"].bool,
  isSmall: _propTypes["default"].bool,
  title: _propTypes["default"].string,
  value: _propTypes["default"].string,
  iconLeft: _propTypes["default"].string
});
(0, _defineProperty2["default"])(DropdownButton, "defaultProps", {
  className: "",
  disabled: false,
  isOpen: false,
  isSmall: true,
  title: "",
  value: "",
  iconLeft: ""
});
