// @gooddata/goodstrap/lib/@next/List/DateDatasetsListItem.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.DateDatasetsListItem = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactIntl = require("react-intl");

var _classnames = _interopRequireDefault(require("classnames"));

var _jsUtils = require("@gooddata/js-utils");

var _ShortenedText = _interopRequireDefault(require("../../core/ShortenedText"));

// (C) 2007-2020 GoodData Corporation
// @ts-ignore
var DateDatasetsListItem = function DateDatasetsListItem(_ref) {
  var id = _ref.id,
      title = _ref.title,
      isHeader = _ref.isHeader,
      isSelected = _ref.isSelected,
      isUnrelated = _ref.isUnrelated,
      onClick = _ref.onClick;

  if (isHeader) {
    return _react["default"].createElement("div", {
      className: "gd-list-item gd-list-item-header"
    }, _react["default"].createElement(_reactIntl.FormattedMessage, {
      id: title
    }));
  }

  var classeNames = (0, _classnames["default"])("gd-list-item", "gd-list-item-shortened", "s-".concat(id), "s-".concat(_jsUtils.string.simplifyText(title)), {
    "is-selected": isSelected,
    "is-unrelated": isUnrelated
  });
  var tooltipAlignPoints = [{
    align: "cl cr",
    offset: {
      x: -10,
      y: 0
    }
  }, {
    align: "cr cl",
    offset: {
      x: 10,
      y: 0
    }
  }];
  return _react["default"].createElement("div", {
    className: classeNames,
    onClick: onClick
  }, _react["default"].createElement(_ShortenedText["default"], {
    tooltipAlignPoints: tooltipAlignPoints
  }, title));
};

exports.DateDatasetsListItem = DateDatasetsListItem;
