// @gooddata/goodstrap/lib/@next/List/MultiSelectListItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.MultiSelectListItem = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _reactIntl = require("react-intl");

var _classnames = _interopRequireDefault(require("classnames"));

var _jsUtils = require("@gooddata/js-utils");

// (C) 2007-2020 GoodData Corporation
var MultiSelectListItem = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2["default"])(MultiSelectListItem, _PureComponent);

  function MultiSelectListItem() {
    var _this;

    (0, _classCallCheck2["default"])(this, MultiSelectListItem);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(MultiSelectListItem).apply(this, arguments));

    _this.getClassNames = function () {
      var _cx;

      var _this$props = _this.props,
          title = _this$props.title,
          isSelected = _this$props.isSelected;
      return (0, _classnames["default"])((_cx = {
        "gd-list-item": true
      }, (0, _defineProperty2["default"])(_cx, "s-".concat(_jsUtils.string.simplifyText(title)), true), (0, _defineProperty2["default"])(_cx, "has-only-visible", true), (0, _defineProperty2["default"])(_cx, "is-selected", isSelected), _cx));
    };

    _this.renderOnly = function () {
      var onOnly = _this.props.onOnly;
      return _react["default"].createElement("span", {
        className: "gd-list-item-only",
        onClick: function onClick(e) {
          e.stopPropagation();

          if (onOnly) {
            onOnly();
          }
        }
      }, _react["default"].createElement(_reactIntl.FormattedMessage, {
        id: "gs.list.only"
      }));
    };

    return _this;
  }

  (0, _createClass2["default"])(MultiSelectListItem, [{
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          title = _this$props2.title,
          onClick = _this$props2.onClick,
          onMouseOver = _this$props2.onMouseOver,
          onMouseOut = _this$props2.onMouseOut,
          isSelected = _this$props2.isSelected;
      return _react["default"].createElement("div", {
        className: this.getClassNames(),
        onClick: onClick,
        onMouseOver: onMouseOver,
        onMouseOut: onMouseOut
      }, _react["default"].createElement("label", {
        className: "input-checkbox-label"
      }, _react["default"].createElement("input", {
        type: "checkbox",
        className: "input-checkbox",
        readOnly: true,
        checked: isSelected
      }), _react["default"].createElement("span", {
        className: "input-label-text"
      }, title)), this.renderOnly());
    }
  }]);
  return MultiSelectListItem;
}(_react.PureComponent);

exports.MultiSelectListItem = MultiSelectListItem;
