// @gooddata/goodstrap/lib/Messages/MessagesBase.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactTransitionGroup = require("react-transition-group");

var _classnames = _interopRequireDefault(require("classnames"));

var _Message = _interopRequireDefault(require("./Message"));

// (C) 2007-2020 GoodData Corporation
var MessagesBase = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(MessagesBase, _Component);

  function MessagesBase(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, MessagesBase);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(MessagesBase).call(this, props));
    _this.state = {
      shouldShowMore: true
    };
    _this.handleShowMore = _this.handleShowMore.bind((0, _assertThisInitialized2["default"])(_this));
    _this.handleMessageClose = _this.handleMessageClose.bind((0, _assertThisInitialized2["default"])(_this));
    return _this;
  }

  (0, _createClass2["default"])(MessagesBase, [{
    key: "handleMessageClose",
    value: function handleMessageClose(messageId) {
      this.setState({
        shouldShowMore: true
      });
      this.props.onMessageClose(messageId);
    }
  }, {
    key: "handleShowMore",
    value: function handleShowMore(e) {
      e.preventDefault();
      this.setState({
        shouldShowMore: !this.state.shouldShowMore
      });
    }
  }, {
    key: "renderMessageWithShowMore",
    value: function renderMessageWithShowMore(message) {
      /* eslint-disable react/no-danger */
      var showMore = message.showMore,
          showLess = message.showLess,
          errorDetail = message.errorDetail,
          text = message.text,
          type = message.type;
      var shouldShowMore = this.state.shouldShowMore;
      var contentClassNames = (0, _classnames["default"])("gd-message-text-content", "s-message-text-content", type, {
        off: shouldShowMore,
        on: !shouldShowMore
      });
      var showMoreLinkClassNames = (0, _classnames["default"])("gd-message-text-showmorelink", "s-message-text-showmorelink", type);
      return _react["default"].createElement("div", {
        className: "gd-message-text-showmore"
      }, _react["default"].createElement("p", {
        className: "gd-message-text-header"
      }, _react["default"].createElement("span", {
        className: "s-message-text-header-value",
        dangerouslySetInnerHTML: {
          __html: text
        }
      }), _react["default"].createElement("span", {
        className: showMoreLinkClassNames,
        onClick: this.handleShowMore
      }, shouldShowMore ? showMore : showLess)), _react["default"].createElement("div", {
        className: contentClassNames
      }, errorDetail));
    }
  }, {
    key: "renderMessage",
    value: function renderMessage(text) {
      /* eslint-disable react/no-danger */
      return _react["default"].createElement("div", {
        className: "s-message-text-header-value",
        dangerouslySetInnerHTML: {
          __html: text
        }
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var messages = this.props.messages.map(function (message) {
        /* eslint-disable react/no-danger */
        var id = message.id,
            component = message.component,
            showMore = message.showMore,
            text = message.text,
            type = message.type;
        return _react["default"].createElement("div", {
          key: id
        }, _react["default"].createElement(_Message["default"], {
          className: "gd-message-overlay gd-message-overlay".concat(showMore && "-custom"),
          type: type,
          onClose: function onClose() {
            _this2.handleMessageClose(id);
          }
        }, component || (showMore ? _this2.renderMessageWithShowMore(message) : _this2.renderMessage(text))));
        /* eslint-enable react/no-danger */
      });
      return _react["default"].createElement("div", {
        className: "gd-messages"
      }, _react["default"].createElement(_reactTransitionGroup.CSSTransitionGroup, {
        transitionName: "gd-message",
        transitionEnterTimeout: 220,
        transitionLeaveTimeout: 220
      }, messages));
    }
  }]);
  return MessagesBase;
}(_react.Component);

exports["default"] = MessagesBase;
(0, _defineProperty2["default"])(MessagesBase, "propTypes", {
  messages: _propTypes["default"].arrayOf(_propTypes["default"].shape(_Message["default"].propTypes)),
  onMessageClose: _propTypes["default"].func
});
(0, _defineProperty2["default"])(MessagesBase, "defaultProps", {
  messages: [],
  onMessageClose: function onMessageClose() {}
});
