// @gooddata/goodstrap/lib/Messages/Message.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

// (C) 2007-2020 GoodData Corporation
var Message = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(Message, _Component);

  function Message() {
    (0, _classCallCheck2["default"])(this, Message);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(Message).apply(this, arguments));
  }

  (0, _createClass2["default"])(Message, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          type = _this$props.type,
          contrast = _this$props.contrast,
          intensive = _this$props.intensive,
          onClose = _this$props.onClose,
          children = _this$props.children;
      var classes = (0, _classnames["default"])("gd-message", "s-message", (0, _defineProperty2["default"])({
        success: type === "success",
        progress: type === "progress",
        error: type === "error",
        warning: type === "warning",
        contrast: contrast,
        intensive: intensive
      }, this.props.className, !!this.props.className));
      /* eslint-disable jsx-a11y/anchor-is-valid, jsx-a11y/anchor-has-content */

      return _react["default"].createElement("div", {
        className: classes
      }, _react["default"].createElement("div", {
        className: "gd-message-text"
      }, children, onClose && _react["default"].createElement("div", {
        className: "gd-message-dismiss-container"
      }, _react["default"].createElement("a", {
        className: "gd-message-dismiss icon-cross",
        onClick: onClose
      }))));
      /* eslint-enable jsx-a11y/anchor-is-valid, jsx-a11y/anchor-has-content */
    }
  }]);
  return Message;
}(_react.Component);

exports["default"] = Message;
(0, _defineProperty2["default"])(Message, "propTypes", {
  children: _propTypes["default"].node,
  className: _propTypes["default"].string,
  contrast: _propTypes["default"].bool,
  intensive: _propTypes["default"].bool,
  onClose: _propTypes["default"].func,
  type: _propTypes["default"].oneOf(["success", "progress", "error", "warning"]).isRequired
});
(0, _defineProperty2["default"])(Message, "defaultProps", {
  children: false,
  className: "",
  contrast: false,
  intensive: false,
  onClose: null
});
