// @gooddata/react-components/dist/components/visualizations/chart/highcharts/areaConfiguration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// (C) 2007-2018 GoodData Corporation
var cloneDeep = require("lodash/cloneDeep");
var LINE_WIDTH = 3;
var AREA_TEMPLATE = {
    chart: {
        type: "area",
    },
    plotOptions: {
        area: {
            lineWidth: 1,
        },
        series: {
            marker: {
                symbol: "circle",
                radius: 4.5,
            },
            lineWidth: LINE_WIDTH,
            fillOpacity: 0.6,
            states: {
                hover: {
                    lineWidth: LINE_WIDTH + 1,
                },
            },
        },
        column: {
            dataLabels: {},
        },
    },
    xAxis: [
        {
            categories: [],
        },
    ],
    yAxis: [
        {
            stackLabels: {
                enabled: false,
            },
        },
    ],
};
function getAreaConfiguration() {
    return cloneDeep(AREA_TEMPLATE);
}
exports.getAreaConfiguration = getAreaConfiguration;
