// @gooddata/react-components/dist/components/visualizations/chart/highcharts/bubbleConfiguration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// (C) 2007-2018 GoodData Corporation
var cloneDeep = require("lodash/cloneDeep");
var BUBBLE_TEMPLATE = {
    chart: {
        type: "bubble",
    },
    plotOptions: {
        bubble: {
            stickyTracking: false,
            marker: {
                symbol: "circle",
                radius: 5,
                states: {
                    hover: {
                        enabled: true,
                        lineColor: "white",
                    },
                },
            },
            states: {
                hover: {
                    marker: {
                        enabled: false,
                    },
                },
            },
            dataLabels: {
                enabled: false,
                allowOverlap: false,
            },
        },
    },
    xAxis: [
        {
            startOnTick: true,
        },
    ],
    series: {
        states: {
            hover: {
                enabled: false,
            },
        },
    },
    legend: {
        enabled: false,
    },
};
function getBubbleConfiguration() {
    return cloneDeep(BUBBLE_TEMPLATE);
}
exports.getBubbleConfiguration = getBubbleConfiguration;
