// @gooddata/react-components/dist/components/visualizations/chart/highcharts/bulletConfiguration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// (C) 2007-2020 GoodData Corporation
var cloneDeep = require("lodash/cloneDeep");
var commonConfiguration_1 = require("./commonConfiguration");
var BULLET_TEMPLATE = {
    chart: {
        type: "bar",
    },
    plotOptions: {
        bar: {
            maxPointWidth: commonConfiguration_1.MAX_POINT_WIDTH,
            dataLabels: {
                enabled: false,
            },
        },
        bullet: {
            maxPointWidth: commonConfiguration_1.MAX_POINT_WIDTH,
            tooltip: {
                followPointer: true,
            },
            targetOptions: {
                borderWidth: 0,
            },
        },
        series: {
            states: {
                hover: {
                    enabled: false,
                },
            },
            grouping: false,
            borderWidth: 0,
        },
    },
    yAxis: [
        {
            stackLabels: {
                enabled: false,
            },
        },
    ],
};
function getBulletConfiguration() {
    return cloneDeep(BULLET_TEMPLATE);
}
exports.getBulletConfiguration = getBulletConfiguration;
