// @gooddata/react-components/dist/components/visualizations/chart/highcharts/lineConfiguration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// (C) 2007-2018 GoodData Corporation
var cloneDeep = require("lodash/cloneDeep");
exports.LINE_WIDTH = 3;
var LINE_TEMPLATE = {
    chart: {
        type: "line",
    },
    plotOptions: {
        series: {
            marker: {
                symbol: "circle",
                radius: 4.5,
            },
            lineWidth: exports.LINE_WIDTH,
            fillOpacity: 0.3,
            states: {
                hover: {
                    lineWidth: exports.LINE_WIDTH + 1,
                },
                inactive: {
                    opacity: 1,
                },
            },
        },
        column: {
            dataLabels: {},
        },
    },
    xAxis: [
        {
            categories: [],
        },
    ],
    yAxis: [
        {
            stackLabels: {
                enabled: false,
            },
        },
    ],
};
function getLineConfiguration() {
    return cloneDeep(LINE_TEMPLATE);
}
exports.getLineConfiguration = getLineConfiguration;
