// @gooddata/react-components/dist/components/visualizations/chart/highcharts/pieConfiguration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// (C) 2007-2018 GoodData Corporation
var cloneDeep = require("lodash/cloneDeep");
var get = require("lodash/get");
var helpers_1 = require("./helpers");
var PIE_TEMPLATE = {
    chart: {
        type: "pie",
        events: {
            load: function () {
                this.series[0].update({
                    dataLabels: {
                        distance: -(get(this, "series.0.points.0.shapeArgs.r", 30) / 3),
                    },
                });
                helpers_1.alignChart(this);
            },
        },
    },
    plotOptions: {
        pie: {
            size: "100%",
            allowPointSelect: false,
            dataLabels: {
                enabled: false,
            },
            showInLegend: true,
        },
    },
    legend: {
        enabled: false,
    },
};
function getPieConfiguration() {
    return cloneDeep(PIE_TEMPLATE);
}
exports.getPieConfiguration = getPieConfiguration;
