// @gooddata/react-components/dist/components/visualizations/chart/highcharts/scatterConfiguration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// (C) 2007-2018 GoodData Corporation
var cloneDeep = require("lodash/cloneDeep");
exports.LINE_WIDTH = 3;
var SCATTER_TEMPLATE = {
    chart: {
        type: "scatter",
    },
    plotOptions: {
        scatter: {
            marker: {
                symbol: "circle",
                radius: 5,
                states: {
                    hover: {
                        enabled: true,
                        lineColor: "white",
                    },
                },
            },
            states: {
                hover: {
                    marker: {
                        enabled: false,
                    },
                },
            },
        },
    },
    xAxis: [
        {
            startOnTick: true,
        },
    ],
    series: {
        lineWidth: exports.LINE_WIDTH,
        fillOpacity: 0.3,
        states: {
            hover: {
                lineWidth: exports.LINE_WIDTH + 1,
            },
        },
    },
    legend: {
        enabled: false,
    },
};
function getScatterConfiguration() {
    return cloneDeep(SCATTER_TEMPLATE);
}
exports.getScatterConfiguration = getScatterConfiguration;
