// @gooddata/react-components/dist/internal/components/configurationControls/colors/colorDropdown/ColorOverlay.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
// (C) 2019 GoodData Corporation
var React = require("react");
var Overlay_1 = require("@gooddata/goodstrap/lib/core/Overlay");
var DropdownVersionType;
(function (DropdownVersionType) {
    DropdownVersionType[DropdownVersionType["ColorPalette"] = 0] = "ColorPalette";
    DropdownVersionType[DropdownVersionType["ColorPicker"] = 1] = "ColorPicker";
})(DropdownVersionType = exports.DropdownVersionType || (exports.DropdownVersionType = {}));
var ALIGN_POINTS_COLOR_PALETTE_PICKER = [
    {
        align: "bl tl",
        offset: {
            x: 0,
            y: 2,
        },
    },
    {
        align: "tl bl",
        offset: {
            x: 0,
            y: 2,
        },
    },
];
var ALIGN_POINTS_CUSTOM_COLOR_PICKER = [
    {
        align: "cr cl",
        offset: {
            x: 5,
            y: 0,
        },
    },
    {
        align: "br bl",
        offset: {
            x: 5,
            y: 100,
        },
    },
    {
        align: "br bl",
        offset: {
            x: 5,
            y: 0,
        },
    },
];
var ColorOverlay = /** @class */ (function (_super) {
    __extends(ColorOverlay, _super);
    function ColorOverlay() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.stopScrollingPropagation = function () {
            document.body.addEventListener("wheel", _this.stopPropagation);
        };
        _this.startScrollingPropagation = function () {
            document.body.removeEventListener("wheel", _this.stopPropagation);
        };
        _this.stopPropagation = function (e) {
            e.stopPropagation();
        };
        _this.onClose = function () {
            _this.props.onClose();
        };
        return _this;
    }
    ColorOverlay.prototype.componentWillUnmount = function () {
        this.startScrollingPropagation();
    };
    ColorOverlay.prototype.render = function () {
        return (React.createElement(Overlay_1.default, { alignTo: this.props.alignTo, onClose: this.onClose, alignPoints: this.getAlignPoints(), closeOnOutsideClick: true, closeOnParentScroll: true, closeOnMouseDrag: true },
            React.createElement("div", { onMouseOver: this.stopScrollingPropagation, onMouseOut: this.startScrollingPropagation }, this.props.children)));
    };
    ColorOverlay.prototype.getAlignPoints = function () {
        if (this.props.dropdownVersion === DropdownVersionType.ColorPalette) {
            return ALIGN_POINTS_COLOR_PALETTE_PICKER;
        }
        return ALIGN_POINTS_CUSTOM_COLOR_PICKER;
    };
    return ColorOverlay;
}(React.PureComponent));
exports.default = ColorOverlay;
