// @gooddata/react-components/dist/internal/components/configurationControls/colors/colorDropdown/ColorDropdown.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
// (C) 2019-2020 GoodData Corporation
var React = require("react");
var ColorPicker_1 = require("@gooddata/goodstrap/lib/ColorPicker/ColorPicker");
var react_intl_1 = require("react-intl");
var gooddata_js_1 = require("@gooddata/gooddata-js");
var uuid_1 = require("uuid");
var ColorOverlay_1 = require("./ColorOverlay");
var ColorPalette_1 = require("./ColorPalette");
var CustomColorButton_1 = require("./CustomColorButton");
var IconPosition;
(function (IconPosition) {
    IconPosition[IconPosition["Down"] = 0] = "Down";
    IconPosition[IconPosition["Right"] = 1] = "Right";
})(IconPosition = exports.IconPosition || (exports.IconPosition = {}));
var COLOR_FOR_UNKNOWN_ITEM = {
    r: 255,
    g: 0,
    b: 0,
};
var ColorDropdown = /** @class */ (function (_super) {
    __extends(ColorDropdown, _super);
    function ColorDropdown(props) {
        var _this = _super.call(this, props) || this;
        _this.onColorPickerSubmit = function (color) {
            var item = {
                type: "rgb",
                value: color,
            };
            _this.onColorSelected(item);
        };
        _this.onColorPickerCancel = function () {
            _this.setState({ dropdownVersion: ColorOverlay_1.DropdownVersionType.ColorPalette });
        };
        _this.onCustomColorButtonClick = function () {
            _this.setState({ dropdownVersion: ColorOverlay_1.DropdownVersionType.ColorPicker });
        };
        _this.onClose = function () {
            _this.setState({ isDropdownOpen: false, dropdownVersion: ColorOverlay_1.DropdownVersionType.ColorPalette });
        };
        _this.onDropdownButtonClick = function () {
            _this.toggleDropdown();
        };
        _this.onColorSelected = function (color) {
            _this.setState({
                isDropdownOpen: false,
                dropdownVersion: ColorOverlay_1.DropdownVersionType.ColorPalette,
            });
            setTimeout(function () {
                _this.props.onColorSelected(color);
            }, 100);
        };
        _this.id = uuid_1.v4();
        _this.state = {
            isDropdownOpen: false,
            dropdownVersion: ColorOverlay_1.DropdownVersionType.ColorPalette,
        };
        return _this;
    }
    ColorDropdown.prototype.render = function () {
        return (React.createElement(React.Fragment, null,
            React.createElement("div", { className: this.getClassName(), onClick: this.onDropdownButtonClick }, this.setupDropdownChild()),
            this.state.isDropdownOpen && (React.createElement(ColorOverlay_1.default, { alignTo: "." + this.getClassName(), onClose: this.onClose, dropdownVersion: this.state.dropdownVersion, key: this.state.dropdownVersion },
                React.createElement("div", { className: "overlay dropdown-body" }, this.isColorPaletteContent()
                    ? this.renderColorPaletteContent()
                    : this.renderColorPickerContent())))));
    };
    ColorDropdown.prototype.setupDropdownChild = function () {
        var childProps = {
            isSelected: this.state.isDropdownOpen,
            position: this.getIconPosition(),
        };
        return React.cloneElement(this.props.children, childProps);
    };
    ColorDropdown.prototype.getIconPosition = function () {
        return this.state.dropdownVersion === ColorOverlay_1.DropdownVersionType.ColorPalette
            ? IconPosition.Down
            : IconPosition.Right;
    };
    ColorDropdown.prototype.isColorPaletteContent = function () {
        return this.state.dropdownVersion === ColorOverlay_1.DropdownVersionType.ColorPalette;
    };
    ColorDropdown.prototype.renderColorPaletteContent = function () {
        return (React.createElement("div", { className: "gd-color-drop-down" },
            React.createElement(ColorPalette_1.default, { selectedColorGuid: this.getSelectedGuidFromColorItem(), colorPalette: this.props.colorPalette, onColorSelected: this.onColorSelected }),
            this.props.showCustomPicker && React.createElement(CustomColorButton_1.default, { onClick: this.onCustomColorButtonClick })));
    };
    ColorDropdown.prototype.getSelectedGuidFromColorItem = function () {
        if (gooddata_js_1.TypeGuards.isGuidColorItem(this.props.selectedColorItem)) {
            return this.props.selectedColorItem.value;
        }
        return null;
    };
    ColorDropdown.prototype.renderColorPickerContent = function () {
        return (React.createElement(ColorPicker_1.default, { initialRgbColor: this.getSelectedColorFromPalette(), onSubmit: this.onColorPickerSubmit, onCancel: this.onColorPickerCancel, intl: this.props.intl }));
    };
    ColorDropdown.prototype.getSelectedColorFromPalette = function () {
        var _this = this;
        if (gooddata_js_1.TypeGuards.isGuidColorItem(this.props.selectedColorItem)) {
            var selected = this.props.colorPalette.find(function (item) {
                return item.guid === _this.props.selectedColorItem.value;
            });
            if (selected) {
                return selected.fill;
            }
        }
        if (gooddata_js_1.TypeGuards.isRgbColorItem(this.props.selectedColorItem)) {
            return this.props.selectedColorItem.value;
        }
        return COLOR_FOR_UNKNOWN_ITEM;
    };
    ColorDropdown.prototype.getClassName = function () {
        return "s-color-drop-down-button-" + this.id;
    };
    ColorDropdown.prototype.toggleDropdown = function () {
        this.setState({
            isDropdownOpen: !this.state.isDropdownOpen,
            dropdownVersion: ColorOverlay_1.DropdownVersionType.ColorPalette,
        });
    };
    return ColorDropdown;
}(React.PureComponent));
exports.default = react_intl_1.injectIntl(ColorDropdown);
