// @gooddata/react-components/dist/internal/constants/axis.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var _a, _b;
// (C) 2019 GoodData Corporation
var visualizationTypes_1 = require("../../constants/visualizationTypes");
exports.AXIS = {
    PRIMARY: "primary",
    SECONDARY: "secondary",
    DUAL: "dual",
};
var AXIS_NAME;
(function (AXIS_NAME) {
    AXIS_NAME["X"] = "xaxis";
    AXIS_NAME["Y"] = "yaxis";
    AXIS_NAME["SECONDARY_X"] = "secondary_xaxis";
    AXIS_NAME["SECONDARY_Y"] = "secondary_yaxis";
})(AXIS_NAME = exports.AXIS_NAME || (exports.AXIS_NAME = {}));
exports.DUAL_AXES_SUPPORTED_CHARTS = [
    visualizationTypes_1.VisualizationTypes.COLUMN,
    visualizationTypes_1.VisualizationTypes.BAR,
    visualizationTypes_1.VisualizationTypes.LINE,
    visualizationTypes_1.VisualizationTypes.COMBO,
];
var BASE_X_AXIS = {
    name: AXIS_NAME.X,
    title: "properties.xaxis.title",
    subtitle: "",
    primary: false,
};
var BASE_Y_AXIS = {
    name: AXIS_NAME.Y,
    title: "properties.yaxis.title",
    subtitle: "",
    primary: true,
};
var BASE_SECONDARY_Y_AXIS = __assign({}, BASE_Y_AXIS, { name: AXIS_NAME.SECONDARY_Y });
var BAR_X_AXIS = __assign({}, BASE_X_AXIS, { primary: true });
var BAR_SECONDARY_X_AXIS = __assign({}, BASE_X_AXIS, { name: AXIS_NAME.SECONDARY_X, primary: true });
var BAR_Y_AXIS = __assign({}, BASE_Y_AXIS, { primary: false });
exports.BASE_CHART_AXIS_CONFIG = (_a = {},
    _a[exports.AXIS.PRIMARY] = [BASE_X_AXIS, BASE_Y_AXIS],
    _a[exports.AXIS.SECONDARY] = [BASE_X_AXIS, BASE_SECONDARY_Y_AXIS],
    _a[exports.AXIS.DUAL] = [
        BASE_X_AXIS,
        __assign({}, BASE_Y_AXIS, { subtitle: "properties.axis.left" }),
        __assign({}, BASE_SECONDARY_Y_AXIS, { subtitle: "properties.axis.right" }),
    ],
    _a);
exports.BAR_CHART_AXIS_CONFIG = (_b = {},
    _b[exports.AXIS.PRIMARY] = [BAR_X_AXIS, BAR_Y_AXIS],
    _b[exports.AXIS.SECONDARY] = [BAR_SECONDARY_X_AXIS, BAR_Y_AXIS],
    _b[exports.AXIS.DUAL] = [
        __assign({}, BAR_SECONDARY_X_AXIS, { subtitle: "properties.axis.top" }),
        __assign({}, BAR_X_AXIS, { subtitle: "properties.axis.bottom" }),
        BAR_Y_AXIS,
    ],
    _b);
