// @gooddata/react-components/dist/internal/constants/supportedProperties.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _a, _b, _c, _d;
// (C) 2019-2020 GoodData Corporation
var axis_1 = require("./axis");
var BASE_PROPERTIES = ["dataLabels.visible", "grid", "legend", "colorMapping"];
var BASE_X_AXIS_PROPERTIES = [
    "xaxis.rotation",
    "xaxis.labelsEnabled",
    "xaxis.visible",
    "xaxis.name.visible",
    "xaxis.name.position",
];
var BASE_Y_AXIS_PROPERTIES = [
    "yaxis.rotation",
    "yaxis.labelsEnabled",
    "yaxis.visible",
    "yaxis.name.visible",
    "yaxis.name.position",
];
var BASE_PRIMARY_AXIS_PROPERTIES = BASE_Y_AXIS_PROPERTIES.concat(["yaxis.min", "yaxis.max"]);
var BAR_PRIMARY_AXIS_PROPERTIES = BASE_X_AXIS_PROPERTIES.concat(["xaxis.min", "xaxis.max"]);
var BASE_SECONDARY_AXIS_PROPERTIES = [
    "secondary_yaxis.rotation",
    "secondary_yaxis.labelsEnabled",
    "secondary_yaxis.visible",
    "secondary_yaxis.min",
    "secondary_yaxis.max",
    "secondary_yaxis.measures",
    "secondary_yaxis.name.visible",
    "secondary_yaxis.name.position",
];
var BAR_SECONDARY_AXIS_PROPERTIES = [
    "secondary_xaxis.rotation",
    "secondary_xaxis.labelsEnabled",
    "secondary_xaxis.visible",
    "secondary_xaxis.min",
    "secondary_xaxis.max",
    "secondary_xaxis.measures",
    "secondary_xaxis.name.visible",
    "secondary_xaxis.name.position",
];
exports.OPTIONAL_STACKING_PROPERTIES = ["stackMeasures", "stackMeasuresToPercent"];
exports.CHART_TYPE_PROPERTIES = ["primaryChartType", "secondaryChartType", "dualAxis"];
exports.BASE_CHART_SUPPORTED_PROPERTIES = BASE_PROPERTIES.concat(BASE_X_AXIS_PROPERTIES, BASE_PRIMARY_AXIS_PROPERTIES);
exports.AREA_CHART_SUPPORTED_PROPERTIES = exports.BASE_CHART_SUPPORTED_PROPERTIES.concat(exports.OPTIONAL_STACKING_PROPERTIES);
exports.COLUMN_CHART_SUPPORTED_PROPERTIES = (_a = {},
    _a[axis_1.AXIS.PRIMARY] = exports.BASE_CHART_SUPPORTED_PROPERTIES.concat(exports.OPTIONAL_STACKING_PROPERTIES),
    _a[axis_1.AXIS.SECONDARY] = BASE_PROPERTIES.concat(BASE_X_AXIS_PROPERTIES, BASE_SECONDARY_AXIS_PROPERTIES, exports.OPTIONAL_STACKING_PROPERTIES),
    _a[axis_1.AXIS.DUAL] = exports.BASE_CHART_SUPPORTED_PROPERTIES.concat(BASE_SECONDARY_AXIS_PROPERTIES, exports.OPTIONAL_STACKING_PROPERTIES),
    _a);
exports.LINE_CHART_SUPPORTED_PROPERTIES = (_b = {},
    _b[axis_1.AXIS.PRIMARY] = exports.BASE_CHART_SUPPORTED_PROPERTIES,
    _b[axis_1.AXIS.SECONDARY] = BASE_PROPERTIES.concat(BASE_X_AXIS_PROPERTIES, BASE_SECONDARY_AXIS_PROPERTIES),
    _b[axis_1.AXIS.DUAL] = exports.BASE_CHART_SUPPORTED_PROPERTIES.concat(BASE_SECONDARY_AXIS_PROPERTIES),
    _b);
exports.BAR_CHART_SUPPORTED_PROPERTIES = (_c = {},
    _c[axis_1.AXIS.PRIMARY] = BASE_PROPERTIES.concat(BASE_Y_AXIS_PROPERTIES, BAR_PRIMARY_AXIS_PROPERTIES, exports.OPTIONAL_STACKING_PROPERTIES),
    _c[axis_1.AXIS.SECONDARY] = BASE_PROPERTIES.concat(BASE_Y_AXIS_PROPERTIES, BAR_SECONDARY_AXIS_PROPERTIES, exports.OPTIONAL_STACKING_PROPERTIES),
    _c[axis_1.AXIS.DUAL] = BASE_PROPERTIES.concat(BASE_Y_AXIS_PROPERTIES, BAR_PRIMARY_AXIS_PROPERTIES, BAR_SECONDARY_AXIS_PROPERTIES, exports.OPTIONAL_STACKING_PROPERTIES),
    _c);
exports.COMBO_CHART_SUPPORTED_PROPERTIES = (_d = {},
    _d[axis_1.AXIS.PRIMARY] = exports.BASE_CHART_SUPPORTED_PROPERTIES.concat(exports.CHART_TYPE_PROPERTIES, exports.OPTIONAL_STACKING_PROPERTIES),
    _d[axis_1.AXIS.SECONDARY] = BASE_PROPERTIES.concat(BASE_X_AXIS_PROPERTIES, BASE_SECONDARY_AXIS_PROPERTIES, exports.CHART_TYPE_PROPERTIES, exports.OPTIONAL_STACKING_PROPERTIES),
    _d[axis_1.AXIS.DUAL] = exports.BASE_CHART_SUPPORTED_PROPERTIES.concat(BASE_SECONDARY_AXIS_PROPERTIES, exports.CHART_TYPE_PROPERTIES, exports.OPTIONAL_STACKING_PROPERTIES),
    _d);
exports.BUBBLE_CHART_SUPPORTED_PROPERTIES = [
    "dataLabels.visible",
    "grid",
    "xaxis.rotation",
    "xaxis.labelsEnabled",
    "xaxis.visible",
    "xaxis.min",
    "xaxis.max",
    "xaxis.name.visible",
    "xaxis.name.position",
    "yaxis.labelsEnabled",
    "yaxis.rotation",
    "yaxis.visible",
    "yaxis.min",
    "yaxis.max",
    "yaxis.name.visible",
    "yaxis.name.position",
    "legend",
    "colorMapping",
];
exports.SCATTERPLOT_SUPPORTED_PROPERTIES = [
    "dataLabels.visible",
    "grid",
    "xaxis.rotation",
    "xaxis.labelsEnabled",
    "xaxis.visible",
    "xaxis.min",
    "xaxis.max",
    "xaxis.name.visible",
    "xaxis.name.position",
    "yaxis.labelsEnabled",
    "yaxis.rotation",
    "yaxis.visible",
    "yaxis.min",
    "yaxis.max",
    "yaxis.name.visible",
    "yaxis.name.position",
    "colorMapping",
];
exports.PIECHART_SUPPORTED_PROPERTIES = [
    "dataLabels.visible",
    "legend",
    "colorMapping",
    "chart.verticalAlign",
];
exports.TREEMAP_SUPPORTED_PROPERTIES = ["dataLabels.visible", "legend", "colorMapping"];
exports.HEATMAP_SUPPORTED_PROPERTIES = [
    "dataLabels.visible",
    "legend",
    "yaxis.rotation",
    "yaxis.labelsEnabled",
    "yaxis.visible",
    "yaxis.name.visible",
    "yaxis.name.position",
    "xaxis.labelsEnabled",
    "xaxis.rotation",
    "xaxis.visible",
    "xaxis.name.visible",
    "xaxis.name.position",
    "colorMapping",
];
exports.BULLET_CHART_SUPPORTED_PROPERTIES = [
    "grid",
    "xaxis.rotation",
    "xaxis.labelsEnabled",
    "xaxis.visible",
    "xaxis.min",
    "xaxis.max",
    "xaxis.name.visible",
    "xaxis.name.position",
    "yaxis.rotation",
    "yaxis.labelsEnabled",
    "yaxis.visible",
    "yaxis.name.visible",
    "yaxis.name.position",
    "legend",
    "colorMapping",
];
exports.GEOPUSHPIN_SUPPORTED_PROPERTIES = [
    "tooltipText",
    "legend",
    "points.groupNearbyPoints",
    "points.minSize",
    "points.maxSize",
    "viewport.area",
    "colorMapping",
];
exports.PIVOT_TABLE_SUPPORTED_PROPERTIES = ["columnWidths"];
