// @gooddata/goodstrap/lib/Dialog/ConfirmDialogBase.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty2 = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty2(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/define-property"));

var _defineProperties = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/define-properties"));

var _getOwnPropertyDescriptors = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptors"));

var _getOwnPropertyDescriptor = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptor"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/get-own-property-symbols"));

var _keys = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/keys"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty3 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Button = _interopRequireDefault(require("../Button/Button"));

var _DialogBase2 = _interopRequireDefault(require("./DialogBase"));

function ownKeys(object, enumerableOnly) { var keys = (0, _keys["default"])(object); if (_getOwnPropertySymbols["default"]) { var symbols = (0, _getOwnPropertySymbols["default"])(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return (0, _getOwnPropertyDescriptor["default"])(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty3["default"])(target, key, source[key]); }); } else if (_getOwnPropertyDescriptors["default"]) { (0, _defineProperties["default"])(target, (0, _getOwnPropertyDescriptors["default"])(source)); } else { ownKeys(Object(source)).forEach(function (key) { (0, _defineProperty2["default"])(target, key, (0, _getOwnPropertyDescriptor["default"])(source, key)); }); } } return target; }

var ConfirmDialogBase = /*#__PURE__*/function (_DialogBase) {
  (0, _inherits2["default"])(ConfirmDialogBase, _DialogBase);

  function ConfirmDialogBase() {
    var _this;

    (0, _classCallCheck2["default"])(this, ConfirmDialogBase);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(ConfirmDialogBase).call(this));
    _this.onKeyDown = _this.onKeyDown.bind((0, _assertThisInitialized2["default"])(_this));
    return _this;
  }

  (0, _createClass2["default"])(ConfirmDialogBase, [{
    key: "render",
    value: function render() {
      var dialogClasses = (0, _classnames["default"])(_objectSpread({
        "gd-confirm": true
      }, this.getDialogClasses()));
      var submitButtonClasses = (0, _classnames["default"])({
        "s-dialog-submit-button": true,
        "gd-button-action": this.props.isPositive,
        "gd-button-negative": !this.props.isPositive
      });
      return (// eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
        _react["default"].createElement("div", {
          tabIndex: "0",
          onKeyDown: this.onKeyDown
        }, _react["default"].createElement("div", {
          className: dialogClasses
        }, this.props.displayCloseButton && this.renderCloseButton(), _react["default"].createElement("div", {
          className: "gd-dialog-header"
        }, _react["default"].createElement("h3", null, this.props.headline)), _react["default"].createElement("div", {
          className: "gd-dialog-content"
        }, this.props.children), _react["default"].createElement("div", {
          className: "gd-dialog-footer"
        }, _react["default"].createElement(_Button["default"], {
          onClick: this.props.onCancel,
          className: "gd-button-secondary s-dialog-cancel-button",
          value: this.props.cancelButtonText
        }), _react["default"].createElement(_Button["default"], {
          onClick: this.props.onSubmit,
          className: submitButtonClasses,
          value: this.props.submitButtonText,
          disabled: this.props.isSubmitDisabled
        }))))
      );
    }
  }]);
  return ConfirmDialogBase;
}(_DialogBase2["default"]);

exports["default"] = ConfirmDialogBase;
(0, _defineProperty3["default"])(ConfirmDialogBase, "propTypes", {
  displayCloseButton: _propTypes["default"].bool,
  onCancel: _propTypes["default"].func,
  onSubmit: _propTypes["default"].func,
  isSubmitDisabled: _propTypes["default"].bool
});
(0, _defineProperty3["default"])(ConfirmDialogBase, "defaultProps", {
  displayCloseButton: true,
  onCancel: function onCancel() {},
  onSubmit: function onSubmit() {},
  isSubmitDisabled: false
});
