// @gooddata/goodstrap/lib/Dialog/ConfirmDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _Overlay = _interopRequireDefault(require("../core/Overlay"));

var _ConfirmDialogBase = _interopRequireDefault(require("./ConfirmDialogBase"));

// (C) 2007-2020 GoodData Corporation
var ReactConfirmDialog = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2["default"])(ReactConfirmDialog, _React$Component);

  function ReactConfirmDialog() {
    (0, _classCallCheck2["default"])(this, ReactConfirmDialog);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(ReactConfirmDialog).apply(this, arguments));
  }

  (0, _createClass2["default"])(ReactConfirmDialog, [{
    key: "render",
    value: function render() {
      return _react["default"].createElement(_Overlay["default"], {
        alignPoints: [{
          align: "cc cc"
        }],
        isModal: true,
        positionType: "fixed"
      }, _react["default"].createElement(_ConfirmDialogBase["default"], this.props));
    }
  }]);
  return ReactConfirmDialog;
}(_react["default"].Component);

var _default = ReactConfirmDialog;
exports["default"] = _default;
