// @gooddata/goodstrap/lib/Dialog/DialogBase.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Button = _interopRequireDefault(require("../Button/Button"));

var _constants = require("../core/constants");

// (C) 2007-2020 GoodData Corporation
// eslint-disable-next-line import/no-unresolved, import/extensions
var checkKeyHandler = function checkKeyHandler(e, keyCode, handler) {
  if (e.keyCode === keyCode && handler) {
    e.preventDefault();
    e.stopPropagation();
    handler();
  }
};

var isTextInputComponent = function isTextInputComponent(e) {
  var _e$target = e.target,
      tagName = _e$target.tagName,
      type = _e$target.type;
  var tagNameInLowercase = tagName.toLowerCase();
  var typeInLowercase = type ? type.toLowerCase() : "";
  return tagNameInLowercase === "textarea" || tagNameInLowercase === "input" && typeInLowercase === "text";
};

var DialogBase = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(DialogBase, _Component);

  function DialogBase() {
    var _this;

    (0, _classCallCheck2["default"])(this, DialogBase);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(DialogBase).call(this));
    _this.onKeyDown = _this.onKeyDown.bind((0, _assertThisInitialized2["default"])(_this));
    return _this;
  }

  (0, _createClass2["default"])(DialogBase, [{
    key: "onKeyDown",
    value: function onKeyDown(e) {
      var _this$props = this.props,
          submitOnEnterKey = _this$props.submitOnEnterKey,
          onCancel = _this$props.onCancel,
          onSubmit = _this$props.onSubmit; // don't call onSubmit when pressing enter key on input fields

      var isEnterKeyDownOnInputField = e.keyCode === _constants.KEY_CODE_ENTER && isTextInputComponent(e);

      if (submitOnEnterKey === false && isEnterKeyDownOnInputField) {
        return;
      }

      checkKeyHandler(e, _constants.KEY_CODE_ENTER, onSubmit);
      checkKeyHandler(e, _constants.KEY_CODE_ESCAPE, onCancel);
    }
  }, {
    key: "getDialogClasses",
    value: function getDialogClasses() {
      return (0, _defineProperty2["default"])({
        overlay: true,
        "gd-dialog": true
      }, this.props.className, !!this.props.className);
    }
  }, {
    key: "renderCloseButton",
    value: function renderCloseButton() {
      return _react["default"].createElement("div", {
        className: "gd-dialog-close"
      }, _react["default"].createElement(_Button["default"], {
        className: "gd-button-link gd-button-icon-only icon-cross s-dialog-close-button",
        value: "",
        onClick: this.props.onCancel
      }));
    }
  }, {
    key: "render",
    value: function render() {
      var dialogClasses = this.getDialogClasses();
      return (// eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
        _react["default"].createElement("div", {
          tabIndex: "0",
          onKeyDown: this.onKeyDown
        }, _react["default"].createElement("div", {
          className: (0, _classnames["default"])(dialogClasses)
        }, this.props.displayCloseButton && this.renderCloseButton(), this.props.children))
      );
    }
  }]);
  return DialogBase;
}(_react.Component);

exports["default"] = DialogBase;
(0, _defineProperty2["default"])(DialogBase, "propTypes", {
  children: _propTypes["default"].node,
  className: _propTypes["default"].string,
  displayCloseButton: _propTypes["default"].bool,
  submitOnEnterKey: _propTypes["default"].bool,
  onCancel: _propTypes["default"].func,
  onSubmit: _propTypes["default"].func
});
(0, _defineProperty2["default"])(DialogBase, "defaultProps", {
  children: false,
  className: "",
  displayCloseButton: false,
  submitOnEnterKey: true,
  onCancel: function onCancel() {},
  onSubmit: null
});
