// @gooddata/goodstrap/lib/Dialog/ExportDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _Overlay = _interopRequireDefault(require("../core/Overlay"));

var _ExportDialogBase = _interopRequireDefault(require("./ExportDialogBase"));

// (C) 2007-2020 GoodData Corporation
var ExportDialog = function ExportDialog(props) {
  var displayCloseButton = props.displayCloseButton,
      isPositive = props.isPositive,
      isSubmitDisabled = props.isSubmitDisabled,
      headline = props.headline,
      cancelButtonText = props.cancelButtonText,
      submitButtonText = props.submitButtonText,
      filterContextText = props.filterContextText,
      filterContextTitle = props.filterContextTitle,
      filterContextVisible = props.filterContextVisible,
      includeFilterContext = props.includeFilterContext,
      mergeHeaders = props.mergeHeaders,
      mergeHeadersDisabled = props.mergeHeadersDisabled,
      mergeHeadersText = props.mergeHeadersText,
      mergeHeadersTitle = props.mergeHeadersTitle,
      onCancel = props.onCancel,
      onSubmit = props.onSubmit;
  return _react["default"].createElement(_Overlay["default"], {
    alignPoints: [{
      align: "cc cc"
    }],
    isModal: true,
    positionType: "fixed"
  }, _react["default"].createElement(_ExportDialogBase["default"], {
    displayCloseButton: displayCloseButton,
    isPositive: isPositive,
    isSubmitDisabled: isSubmitDisabled,
    headline: headline,
    cancelButtonText: cancelButtonText,
    submitButtonText: submitButtonText,
    filterContextText: filterContextText,
    filterContextTitle: filterContextTitle,
    filterContextVisible: filterContextVisible,
    includeFilterContext: includeFilterContext,
    mergeHeaders: mergeHeaders,
    mergeHeadersDisabled: mergeHeadersDisabled,
    mergeHeadersText: mergeHeadersText,
    mergeHeadersTitle: mergeHeadersTitle,
    onCancel: onCancel,
    onSubmit: onSubmit
  }));
};

ExportDialog.propTypes = _ExportDialogBase["default"].propTypes;
ExportDialog.defaultProps = _ExportDialogBase["default"].defaultProps;
var _default = ExportDialog;
exports["default"] = _default;
