// @gooddata/goodstrap/lib/@next/Dropdown/DropdownList.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.DropdownList = exports.DEFAULT_MOBILE_ITEM_HEIGHT = exports.DEFAULT_ITEM_HEIGHT = exports.LOADING_HEIGHT = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/get-own-property-symbols"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _reactIntl = require("react-intl");

var _LoadingMask = _interopRequireDefault(require("../../core/LoadingMask"));

var _Input = _interopRequireDefault(require("../../Form/Input"));

var _NoData = require("../NoData");

var _AutoSize = require("../AutoSize/AutoSize");

var _DropdownTabs = require("./DropdownTabs");

var _List = require("../List/List");

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof _getOwnPropertySymbols["default"] === "function") for (var i = 0, p = (0, _getOwnPropertySymbols["default"])(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0) t[p[i]] = s[p[i]];
  }
  return t;
}; // (C) 2007-2020 GoodData Corporation


var LOADING_HEIGHT = 100;
exports.LOADING_HEIGHT = LOADING_HEIGHT;
var DEFAULT_ITEM_HEIGHT = 28;
exports.DEFAULT_ITEM_HEIGHT = DEFAULT_ITEM_HEIGHT;
var DEFAULT_MOBILE_ITEM_HEIGHT = 40;
exports.DEFAULT_MOBILE_ITEM_HEIGHT = DEFAULT_MOBILE_ITEM_HEIGHT;
var defaultNoData = (0, _reactIntl.injectIntl)(function (_ref) {
  var hasNoMatchingData = _ref.hasNoMatchingData,
      intl = _ref.intl;
  return _react["default"].createElement(_NoData.NoData, {
    hasNoMatchingData: hasNoMatchingData,
    notFoundLabel: intl.formatMessage({
      id: "gs.noData.noMatchingData"
    }),
    noDataLabel: intl.formatMessage({
      id: "gs.noData.noDataAvailable"
    })
  });
});

var DropdownList = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(DropdownList, _Component);

  function DropdownList() {
    (0, _classCallCheck2["default"])(this, DropdownList);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(DropdownList).apply(this, arguments));
  }

  (0, _createClass2["default"])(DropdownList, [{
    key: "render",
    value: function render() {
      var _this = this;

      var _a = this.props,
          _a$className = _a.className,
          className = _a$className === void 0 ? "" : _a$className,
          width = _a.width,
          height = _a.height,
          isMobile = _a.isMobile,
          isLoading = _a.isLoading,
          _a$items = _a.items,
          items = _a$items === void 0 ? [] : _a$items,
          _a$itemsCount = _a.itemsCount,
          itemsCount = _a$itemsCount === void 0 ? items.length : _a$itemsCount,
          _a$itemHeight = _a.itemHeight,
          itemHeight = _a$itemHeight === void 0 ? DEFAULT_ITEM_HEIGHT : _a$itemHeight,
          _a$mobileItemHeight = _a.mobileItemHeight,
          mobileItemHeight = _a$mobileItemHeight === void 0 ? DEFAULT_MOBILE_ITEM_HEIGHT : _a$mobileItemHeight,
          showSearch = _a.showSearch,
          disableAutofocus = _a.disableAutofocus,
          searchString = _a.searchString,
          searchPlaceholder = _a.searchPlaceholder,
          searchFieldSize = _a.searchFieldSize,
          onSearch = _a.onSearch,
          showTabs = _a.showTabs,
          tabs = _a.tabs,
          selectedTabId = _a.selectedTabId,
          onTabSelect = _a.onTabSelect,
          _a$renderNoData = _a.renderNoData,
          renderNoData = _a$renderNoData === void 0 ? defaultNoData : _a$renderNoData,
          listProps = __rest(_a, ["className", "width", "height", "isMobile", "isLoading", "items", "itemsCount", "itemHeight", "mobileItemHeight", "showSearch", "disableAutofocus", "searchString", "searchPlaceholder", "searchFieldSize", "onSearch", "showTabs", "tabs", "selectedTabId", "onTabSelect", "renderNoData"]);

      var hasNoData = !isLoading && itemsCount === 0;
      var hasNoMatchingData = hasNoData && !!searchString;
      var listClassNames = (0, _classnames["default"])("gd-infinite-list", className);
      var searchFieldClassNames = (0, _classnames["default"])({
        "gd-list-searchfield": true,
        "gd-flex-item": true
      });

      var renderFooter = function renderFooter() {
        var _this$props = _this.props,
            footer = _this$props.footer,
            closeDropdown = _this$props.closeDropdown;

        if (footer && closeDropdown && typeof footer === "function") {
          return footer(closeDropdown);
        }

        return footer || false;
      };

      return _react["default"].createElement(_react["default"].Fragment, null, showSearch && _react["default"].createElement(_Input["default"], {
        className: searchFieldClassNames,
        value: searchString,
        onChange: onSearch,
        isSmall: searchFieldSize === "small",
        placeholder: searchPlaceholder,
        clearOnEsc: true,
        isSearch: true,
        autofocus: !disableAutofocus
      }), showTabs && _react["default"].createElement(_DropdownTabs.DropdownTabs, {
        tabs: tabs,
        selectedTabId: selectedTabId,
        onTabSelect: onTabSelect
      }), hasNoData && _react["default"].createElement("div", {
        style: {
          width: isMobile ? "auto" : width
        }
      }, renderNoData({
        hasNoMatchingData: hasNoMatchingData
      })), isLoading && _react["default"].createElement(_LoadingMask["default"], {
        style: {
          width: isMobile ? "100%" : width,
          height: LOADING_HEIGHT
        }
      }), !isLoading && itemsCount > 0 && _react["default"].createElement(_AutoSize.AutoSize, null, function (autoSize) {
        var listWidth = isMobile ? autoSize.width : width;
        var listHeight = isMobile ? autoSize.height : height;
        return _react["default"].createElement(_List.List, (0, _extends2["default"])({
          className: listClassNames,
          width: listWidth,
          height: listHeight,
          items: items,
          itemsCount: itemsCount,
          itemHeight: isMobile ? Math.max(mobileItemHeight, itemHeight) : itemHeight
        }, listProps));
      }), renderFooter());
    }
  }]);
  return DropdownList;
}(_react.Component);

exports.DropdownList = DropdownList;
