// @gooddata/goodstrap/lib/@next/Dropdown/DropdownButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.DropdownButton = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Button = _interopRequireDefault(require("../../Button/Button"));

// (C) 2007-2020 GoodData Corporation
var DropdownButton = function DropdownButton(_ref) {
  var className = _ref.className,
      value = _ref.value,
      _ref$title = _ref.title,
      title = _ref$title === void 0 ? value : _ref$title,
      disabled = _ref.disabled,
      isOpen = _ref.isOpen,
      _ref$isSmall = _ref.isSmall,
      isSmall = _ref$isSmall === void 0 ? true : _ref$isSmall,
      iconLeft = _ref.iconLeft,
      onClick = _ref.onClick;
  var buttonClasses = (0, _classnames["default"])({
    "gd-button-primary": true,
    "gd-button-small": isSmall,
    "button-dropdown": true,
    "is-dropdown-open": isOpen,
    "is-active": isOpen,
    "dropdown-button": true
  }, className);
  return _react["default"].createElement(_Button["default"], {
    title: title,
    className: buttonClasses,
    value: value,
    iconLeft: iconLeft,
    iconRight: isOpen ? "icon-navigateup" : "icon-navigatedown",
    disabled: disabled,
    onClick: onClick
  });
};

exports.DropdownButton = DropdownButton;
