// @gooddata/goodstrap/lib/Bubble/BubbleHoverTrigger.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = exports.HIDE_DELAY = exports.SHOW_DELAY = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _BubbleTrigger2 = _interopRequireDefault(require("./BubbleTrigger"));

// (C) 2007-2020 GoodData Corporation
var SHOW_DELAY = 425;
exports.SHOW_DELAY = SHOW_DELAY;
var HIDE_DELAY = 200;
exports.HIDE_DELAY = HIDE_DELAY;

var BubbleHoverTrigger = /*#__PURE__*/function (_BubbleTrigger) {
  (0, _inherits2["default"])(BubbleHoverTrigger, _BubbleTrigger);

  function BubbleHoverTrigger() {
    (0, _classCallCheck2["default"])(this, BubbleHoverTrigger);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(BubbleHoverTrigger).apply(this, arguments));
  }

  (0, _createClass2["default"])(BubbleHoverTrigger, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.cancelBubbleVisibilityChange();
    }
  }, {
    key: "cancelBubbleVisibilityChange",
    value: function cancelBubbleVisibilityChange() {
      if (this.scheduleId) {
        window.clearTimeout(this.scheduleId);
      }
    }
  }, {
    key: "scheduleBubbleVisibilityChange",
    value: function scheduleBubbleVisibilityChange(visible) {
      var _this = this;

      var delay = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
      this.cancelBubbleVisibilityChange();
      this.scheduleId = window.setTimeout(function () {
        _this.changeBubbleVisibility(visible);

        var hoverHideDelay = _this.props.hoverHideDelay;

        if (visible && hoverHideDelay) {
          _this.scheduleBubbleVisibilityChange(false, hoverHideDelay);
        }
      }, delay);
    }
  }, {
    key: "eventListeners",
    value: function eventListeners() {
      return {
        onMouseEnter: this.scheduleBubbleVisibilityChange.bind(this, true, this.props.showDelay),
        onMouseLeave: this.scheduleBubbleVisibilityChange.bind(this, false, this.props.hideDelay)
      };
    }
  }]);
  return BubbleHoverTrigger;
}(_BubbleTrigger2["default"]);

exports["default"] = BubbleHoverTrigger;
(0, _defineProperty2["default"])(BubbleHoverTrigger, "propTypes", {
  showDelay: _propTypes["default"].number,
  hideDelay: _propTypes["default"].number,
  hoverHideDelay: _propTypes["default"].number,
  eventsOnBubble: _propTypes["default"].bool,
  tagName: _propTypes["default"].string,
  className: _propTypes["default"].string
});
(0, _defineProperty2["default"])(BubbleHoverTrigger, "defaultProps", {
  showDelay: SHOW_DELAY,
  hideDelay: HIDE_DELAY,
  hoverHideDelay: 0,
  eventsOnBubble: false,
  tagName: "span"
});
