// @gooddata/goodstrap/lib/Bubble/BubbleTrigger.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty2 = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty2(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/define-property"));

var _defineProperties = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/define-properties"));

var _getOwnPropertyDescriptors = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptors"));

var _getOwnPropertyDescriptor = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptor"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/get-own-property-symbols"));

var _keys = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/keys"));

var _extends2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty3 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _bindAll = _interopRequireDefault(require("lodash/bindAll"));

var _uniqueId = _interopRequireDefault(require("lodash/uniqueId"));

function ownKeys(object, enumerableOnly) { var keys = (0, _keys["default"])(object); if (_getOwnPropertySymbols["default"]) { var symbols = (0, _getOwnPropertySymbols["default"])(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return (0, _getOwnPropertyDescriptor["default"])(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty3["default"])(target, key, source[key]); }); } else if (_getOwnPropertyDescriptors["default"]) { (0, _defineProperties["default"])(target, (0, _getOwnPropertyDescriptors["default"])(source)); } else { ownKeys(Object(source)).forEach(function (key) { (0, _defineProperty2["default"])(target, key, (0, _getOwnPropertyDescriptor["default"])(source, key)); }); } } return target; }

var BubbleTrigger = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(BubbleTrigger, _Component);

  function BubbleTrigger() {
    var _this;

    (0, _classCallCheck2["default"])(this, BubbleTrigger);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(BubbleTrigger).call(this));
    _this.state = {
      bubbleId: (0, _uniqueId["default"])("bubble-"),
      isBubbleVisible: false
    };
    (0, _bindAll["default"])((0, _assertThisInitialized2["default"])(_this), "onClose");
    return _this;
  }

  (0, _createClass2["default"])(BubbleTrigger, [{
    key: "onClose",
    value: function onClose() {
      this.changeBubbleVisibility(false);
    }
  }, {
    key: "getClassnames",
    value: function getClassnames() {
      var _cx;

      return (0, _classnames["default"])((_cx = {
        "gd-bubble-trigger": true
      }, (0, _defineProperty3["default"])(_cx, this.state.bubbleId, true), (0, _defineProperty3["default"])(_cx, this.props.className, !!this.props.className), _cx));
    }
  }, {
    key: "eventListeners",
    value: function eventListeners() {
      return {};
    }
  }, {
    key: "changeBubbleVisibility",
    value: function changeBubbleVisibility(active) {
      this.setState({
        isBubbleVisible: active
      });
    }
  }, {
    key: "render",
    value: function render() {
      var TagName = this.props.tagName;
      var BubbleElement;
      var WrappedTrigger;

      _react["default"].Children.map(this.props.children, function (child) {
        if (child) {
          if (child.type && child.type.identifier === "Bubble") {
            BubbleElement = child;
          } else {
            WrappedTrigger = child;
          }
        }
      });

      var bubbleProps = _objectSpread({}, this.props.eventsOnBubble ? this.eventListeners() : {}, {
        alignTo: ".".concat(this.state.bubbleId),
        onClose: this.onClose
      });

      var BubbleOverlay = this.state.isBubbleVisible && BubbleElement ? _react["default"].cloneElement(BubbleElement, bubbleProps) : "";
      return _react["default"].createElement(_react["default"].Fragment, null, _react["default"].createElement(TagName, (0, _extends2["default"])({}, this.eventListeners(), {
        className: this.getClassnames()
      }), WrappedTrigger), BubbleOverlay);
    }
  }]);
  return BubbleTrigger;
}(_react.Component);

exports["default"] = BubbleTrigger;
(0, _defineProperty3["default"])(BubbleTrigger, "propTypes", {
  className: _propTypes["default"].string,
  children: _propTypes["default"].node,
  eventsOnBubble: _propTypes["default"].bool,
  tagName: _propTypes["default"].string
});
(0, _defineProperty3["default"])(BubbleTrigger, "defaultProps", {
  className: "",
  children: false,
  eventsOnBubble: false,
  tagName: "span"
});
