// @gooddata/goodstrap/lib/Bubble/Bubble.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty2 = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty2(exports, "__esModule", {
  value: true
});

exports["default"] = exports.Y_SHIFT = exports.X_SHIFT = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/define-property"));

var _defineProperties = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/define-properties"));

var _getOwnPropertyDescriptors = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptors"));

var _getOwnPropertyDescriptor = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptor"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/get-own-property-symbols"));

var _keys = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/keys"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty3 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _keys2 = _interopRequireDefault(require("lodash/keys"));

var _cloneDeep = _interopRequireDefault(require("lodash/cloneDeep"));

var _isEqual = _interopRequireDefault(require("lodash/isEqual"));

var _result = _interopRequireDefault(require("lodash/result"));

var _bindAll = _interopRequireDefault(require("lodash/bindAll"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Overlay = _interopRequireDefault(require("../core/Overlay"));

var _immutable = require("../core/immutable");

function ownKeys(object, enumerableOnly) { var keys = (0, _keys["default"])(object); if (_getOwnPropertySymbols["default"]) { var symbols = (0, _getOwnPropertySymbols["default"])(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return (0, _getOwnPropertyDescriptor["default"])(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty3["default"])(target, key, source[key]); }); } else if (_getOwnPropertyDescriptors["default"]) { (0, _defineProperties["default"])(target, (0, _getOwnPropertyDescriptors["default"])(source)); } else { ownKeys(Object(source)).forEach(function (key) { (0, _defineProperty2["default"])(target, key, (0, _getOwnPropertyDescriptor["default"])(source, key)); }); } } return target; }

var ARROW_DIRECTIONS = {
  ".. cc": "none",
  ".r .l|.. cl": "left",
  ".l .r|.. cr": "right",
  ".. t.": "top",
  ".. b.": "bottom",
  ".. .l": "left",
  ".. .r": "right"
}; // FIXME: hardcoded offsets for Indigo style

var X_SHIFT = 7;
exports.X_SHIFT = X_SHIFT;
var Y_SHIFT = 11; // FIXME: constats are bad, we know :(

exports.Y_SHIFT = Y_SHIFT;
var ARROW_OFFSETS = {
  ".. cc": [0, 0],
  ".. tc": [0, X_SHIFT],
  ".. bc": [0, -X_SHIFT],
  ".. cl": [X_SHIFT, 0],
  ".. cr": [-X_SHIFT, 0],
  ".r tl": [X_SHIFT, -Y_SHIFT],
  ".l tr": [-X_SHIFT, -Y_SHIFT],
  ".r bl": [X_SHIFT, Y_SHIFT],
  ".l br": [-X_SHIFT, Y_SHIFT],
  ".. tl": [-Y_SHIFT, X_SHIFT],
  ".. tr": [Y_SHIFT, X_SHIFT],
  ".. bl": [-Y_SHIFT, -X_SHIFT],
  ".. br": [Y_SHIFT, -X_SHIFT]
};

var Bubble = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(Bubble, _Component);

  // identifier for BubbleTrigger
  function Bubble(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, Bubble);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(Bubble).call(this, props));
    (0, _bindAll["default"])((0, _assertThisInitialized2["default"])(_this), "onAlign");
    _this.arrowOffsets = _objectSpread({}, props.arrowOffsets, {}, ARROW_OFFSETS);
    _this.arrowDirections = _objectSpread({}, props.arrowDirections, {}, ARROW_DIRECTIONS);

    var alignPoints = _this.addOffsetToAlignPoints((0, _cloneDeep["default"])(props.alignPoints));

    _this.state = {
      alignPoints: alignPoints,
      optimalAlignPoints: props.alignPoints[0].align
    };
    return _this;
  }

  (0, _createClass2["default"])(Bubble, [{
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps, nextState) {
      var propsChanged = !(0, _immutable.propsEqual)(this.props, nextProps);
      var alignmentChanged = !(0, _isEqual["default"])(this.state.optimalAlignPoints, nextState.optimalAlignPoints);
      return propsChanged || alignmentChanged;
    }
  }, {
    key: "onAlign",
    value: function onAlign(alignment) {
      this.setState({
        optimalAlignPoints: alignment.align
      });
    }
  }, {
    key: "getClassnames",
    value: function getClassnames() {
      var _cx;

      return (0, _classnames["default"])((_cx = {}, (0, _defineProperty3["default"])(_cx, this.props.className, !!this.props.className), (0, _defineProperty3["default"])(_cx, this.getArrowsClassname(this.state.optimalAlignPoints), true), (0, _defineProperty3["default"])(_cx, "gd-bubble", true), (0, _defineProperty3["default"])(_cx, "bubble", true), _cx));
    }
  }, {
    key: "getArrowsClassname",
    value: function getArrowsClassname(alignPoints) {
      var myAlignPoint = alignPoints.split(" ")[1];
      var direction = this.getArrowDirection(alignPoints);
      return "arrow-".concat(direction, "-direction arrow-").concat(myAlignPoint);
    }
  }, {
    key: "getArrowDirection",
    value: function getArrowDirection(alignPoints) {
      var key = (0, _keys2["default"])(this.arrowDirections).find(function (arrowDirection) {
        return alignPoints.match(arrowDirection);
      });
      return this.arrowDirections[key] || "none";
    }
  }, {
    key: "addOffsetToAlignPoints",
    value: function addOffsetToAlignPoints(alignPoints) {
      var _this2 = this;

      var arrowOffsets = this.arrowOffsets;
      var arrowOffsetsKeys = (0, _keys2["default"])(arrowOffsets);

      var getKey = function getKey(align, re) {
        return align.match(re) !== null;
      };

      return alignPoints.map(function (item) {
        var key = arrowOffsetsKeys.find(getKey.bind(_this2, item.align));
        item.offset = item.offset || {
          x: 0,
          y: 0
        }; // eslint-disable-line no-param-reassign

        item.offset.x += arrowOffsets[key][0]; // eslint-disable-line no-param-reassign

        item.offset.y += arrowOffsets[key][1]; // eslint-disable-line no-param-reassign

        return item;
      }, this);
    }
  }, {
    key: "render",
    value: function render() {
      var arrowStyle = (0, _result["default"])(this.props, "arrowStyle", {});
      return _react["default"].createElement(_Overlay["default"], {
        className: this.props.overlayClassName,
        alignTo: this.props.alignTo,
        onAlign: this.onAlign,
        alignPoints: this.state.alignPoints,
        closeOnParentScroll: true,
        closeOnMouseDrag: true,
        closeOnOutsideClick: this.props.closeOnOutsideClick,
        onClose: this.props.onClose
      }, _react["default"].createElement("div", {
        onMouseEnter: this.props.onMouseEnter,
        onMouseLeave: this.props.onMouseLeave,
        className: this.getClassnames()
      }, _react["default"].createElement("div", {
        className: "bubble-content"
      }, _react["default"].createElement("div", {
        className: "helper"
      }), _react["default"].createElement("div", {
        className: "arrow-position",
        style: arrowStyle
      }, _react["default"].createElement("div", {
        className: "arrow-border"
      }), _react["default"].createElement("div", {
        className: "arrow"
      })), _react["default"].createElement("div", {
        className: "content"
      }, this.props.children))));
    }
  }]);
  return Bubble;
}(_react.Component);

exports["default"] = Bubble;
(0, _defineProperty3["default"])(Bubble, "propTypes", {
  alignPoints: _propTypes["default"].array,
  alignTo: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].object]),
  arrowDirections: _propTypes["default"].object,
  arrowOffsets: _propTypes["default"].object,
  arrowStyle: _propTypes["default"].oneOfType([_propTypes["default"].func, _propTypes["default"].object]),
  children: _propTypes["default"].node.isRequired,
  className: _propTypes["default"].string,
  closeOnOutsideClick: _propTypes["default"].bool,
  onClose: _propTypes["default"].func,
  onMouseEnter: _propTypes["default"].func,
  onMouseLeave: _propTypes["default"].func,
  overlayClassName: _propTypes["default"].string
});
(0, _defineProperty3["default"])(Bubble, "defaultProps", {
  alignPoints: [{
    align: "bl tl"
  }],
  alignTo: "body",
  arrowOffsets: {},
  arrowDirections: {},
  arrowStyle: {},
  className: "bubble-primary",
  closeOnOutsideClick: false,
  onClose: function onClose() {},
  onMouseEnter: function onMouseEnter() {},
  onMouseLeave: function onMouseLeave() {},
  overlayClassName: ""
});
(0, _defineProperty3["default"])(Bubble, "identifier", "Bubble");
