// @gooddata/goodstrap/lib/Bubble/BubbleFocusTrigger.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _BubbleTrigger2 = _interopRequireDefault(require("./BubbleTrigger"));

// (C) 2007-2020 GoodData Corporation
var BubbleFocusTrigger = /*#__PURE__*/function (_BubbleTrigger) {
  (0, _inherits2["default"])(BubbleFocusTrigger, _BubbleTrigger);

  function BubbleFocusTrigger() {
    (0, _classCallCheck2["default"])(this, BubbleFocusTrigger);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(BubbleFocusTrigger).apply(this, arguments));
  }

  (0, _createClass2["default"])(BubbleFocusTrigger, [{
    key: "eventListeners",
    value: function eventListeners() {
      return {
        onFocus: this.changeBubbleVisibility.bind(this, true),
        onBlur: this.changeBubbleVisibility.bind(this, false)
      };
    }
  }]);
  return BubbleFocusTrigger;
}(_BubbleTrigger2["default"]);

exports["default"] = BubbleFocusTrigger;
(0, _defineProperty2["default"])(BubbleFocusTrigger, "propTypes", {
  tagName: _propTypes["default"].string,
  eventsOnBubble: _propTypes["default"].bool,
  className: _propTypes["default"].string
});
(0, _defineProperty2["default"])(BubbleFocusTrigger, "defaultProps", {
  tagName: "span",
  eventsOnBubble: true
});
