// @gooddata/goodstrap/lib/Header/HeaderMenu.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = exports.HeaderMenu = exports.itemPropType = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactIntl = require("react-intl");

var _uniqueId = _interopRequireDefault(require("lodash/uniqueId"));

var _classnames = _interopRequireDefault(require("classnames"));

var _intlShape = require("../core/intlShape");

// (C) 2007-2020 GoodData Corporation
var itemPropType = _propTypes["default"].shape({
  key: _propTypes["default"].string.isRequired,
  href: _propTypes["default"].string,
  isActive: _propTypes["default"].bool,
  className: _propTypes["default"].string,
  target: _propTypes["default"].string,
  onClick: _propTypes["default"].func
});

exports.itemPropType = itemPropType;

var HeaderMenu = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2["default"])(HeaderMenu, _PureComponent);

  function HeaderMenu() {
    (0, _classCallCheck2["default"])(this, HeaderMenu);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(HeaderMenu).apply(this, arguments));
  }

  (0, _createClass2["default"])(HeaderMenu, [{
    key: "getClassNames",
    value: function getClassNames() {
      return (0, _classnames["default"])("gd-header-menu", this.props.className);
    }
  }, {
    key: "renderSection",
    value: function renderSection(items) {
      var _this = this;

      return items.map(function (item) {
        var clickHandler = item.onClick ? item.onClick : function (event) {
          return _this.props.onMenuItemClick(item, event);
        };
        var classNames = (0, _classnames["default"])("gd-header-menu-item", {
          active: item.isActive
        });
        return _react["default"].createElement("li", {
          key: item.key
        }, _react["default"].createElement("a", {
          onClick: clickHandler // eslint-disable-line react/jsx-no-bind
          ,
          href: item.href,
          className: classNames,
          target: item.target,
          rel: item.target === "_blank" ? "noreferrer noopener" : undefined
        }, _react["default"].createElement("span", {
          className: item.className
        }, _react["default"].createElement(_reactIntl.FormattedMessage, {
          id: item.key
        }))));
      });
    }
  }, {
    key: "renderSections",
    value: function renderSections() {
      var _this2 = this;

      return this.props.sections.map(function (items) {
        return _react["default"].createElement("ul", {
          key: (0, _uniqueId["default"])("section-"),
          className: "gd-header-menu-section gd-header-measure"
        }, _this2.renderSection(items));
      });
    }
  }, {
    key: "render",
    value: function render() {
      return _react["default"].createElement("div", {
        className: this.getClassNames()
      }, this.renderSections());
    }
  }]);
  return HeaderMenu;
}(_react.PureComponent);

exports.HeaderMenu = HeaderMenu;
(0, _defineProperty2["default"])(HeaderMenu, "propTypes", {
  className: _propTypes["default"].string,
  intl: _intlShape.intlShape.isRequired,
  onMenuItemClick: _propTypes["default"].func,
  sections: _propTypes["default"].arrayOf(_propTypes["default"].arrayOf(itemPropType))
});
(0, _defineProperty2["default"])(HeaderMenu, "defaultProps", {
  className: "",
  onMenuItemClick: function onMenuItemClick() {},
  sections: []
});

var _default = (0, _reactIntl.injectIntl)(HeaderMenu);

exports["default"] = _default;
