// @gooddata/goodstrap/lib/Header/HeaderAccount.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = exports.HeaderAccount = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _bindAll = _interopRequireDefault(require("lodash/bindAll"));

var _reactIntl = require("react-intl");

var _Overlay = _interopRequireDefault(require("../core/Overlay"));

var _HeaderMenu = require("./HeaderMenu");

// (C) 2007-2020 GoodData Corporation
var HeaderAccount = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2["default"])(HeaderAccount, _PureComponent);

  function HeaderAccount(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, HeaderAccount);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(HeaderAccount).call(this, props));
    (0, _bindAll["default"])((0, _assertThisInitialized2["default"])(_this), "toggleAccountMenu");
    _this.state = {
      isOpen: false
    };
    return _this;
  }

  (0, _createClass2["default"])(HeaderAccount, [{
    key: "getClassNames",
    value: function getClassNames() {
      return (0, _classnames["default"])((0, _defineProperty2["default"])({
        "gd-header-account": true,
        "is-open": this.state.isOpen
      }, this.props.className, !!this.props.className));
    }
  }, {
    key: "getMenuItems",
    value: function getMenuItems() {
      var _this2 = this;

      return this.props.items.map(function (item) {
        return _react["default"].createElement("a", {
          key: item.key,
          href: item.href,
          onClick: function onClick(e) {
            _this2.menuItemClicked(item, e);
          },
          className: "gd-list-item ".concat(item.className)
        }, _react["default"].createElement(_reactIntl.FormattedMessage, {
          id: item.key
        }));
      });
    }
  }, {
    key: "toggleAccountMenu",
    value: function toggleAccountMenu() {
      var isOpen = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : !this.state.isOpen;
      this.setState({
        isOpen: isOpen
      });
    }
  }, {
    key: "menuItemClicked",
    value: function menuItemClicked() {
      var _this$props;

      this.toggleAccountMenu(false);

      (_this$props = this.props).onMenuItemClick.apply(_this$props, arguments);
    }
  }, {
    key: "renderAccountMenu",
    value: function renderAccountMenu() {
      var _this3 = this;

      return this.state.isOpen ? _react["default"].createElement(_Overlay["default"], {
        alignTo: ".gd-header-account",
        alignPoints: [{
          align: "br tr"
        }],
        closeOnOutsideClick: true,
        closeOnMouseDrag: true,
        closeOnParentScroll: true,
        onClose: function onClose() {
          _this3.toggleAccountMenu(false);
        }
      }, _react["default"].createElement("div", {
        className: "gd-dialog gd-dropdown overlay gd-header-account-dropdown"
      }, _react["default"].createElement("div", {
        className: "gd-list small"
      }, this.getMenuItems()))) : false;
    }
  }, {
    key: "render",
    value: function render() {
      return _react["default"].createElement("div", {
        className: this.getClassNames(),
        onClick: this.toggleAccountMenu
      }, _react["default"].createElement("span", {
        className: "gd-header-account-icon icon-user"
      }), _react["default"].createElement("span", {
        className: "gd-header-account-user"
      }, this.props.userName), this.renderAccountMenu());
    }
  }]);
  return HeaderAccount;
}(_react.PureComponent);

exports.HeaderAccount = HeaderAccount;
(0, _defineProperty2["default"])(HeaderAccount, "propTypes", {
  className: _propTypes["default"].string,
  items: _propTypes["default"].arrayOf(_HeaderMenu.itemPropType),
  onMenuItemClick: _propTypes["default"].func.isRequired,
  userName: _propTypes["default"].string
});
(0, _defineProperty2["default"])(HeaderAccount, "defaultProps", {
  className: "",
  items: [],
  userName: ""
});

var _default = (0, _reactIntl.injectIntl)(HeaderAccount);

exports["default"] = _default;
