// @gooddata/goodstrap/lib/Header/HeaderHelp.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = exports.HeaderHelp = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _classnames = _interopRequireDefault(require("classnames"));

var _bindAll = _interopRequireDefault(require("lodash/bindAll"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _reactIntl = require("react-intl");

var _Overlay = _interopRequireDefault(require("../core/Overlay"));

var _HeaderMenu = require("./HeaderMenu");

// (C) 2007-2020 GoodData Corporation
var HeaderHelp = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2["default"])(HeaderHelp, _PureComponent);

  function HeaderHelp(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, HeaderHelp);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(HeaderHelp).call(this, props));
    (0, _bindAll["default"])((0, _assertThisInitialized2["default"])(_this), "toggleHelpMenu");
    _this.state = {
      isOpen: false
    };
    return _this;
  }

  (0, _createClass2["default"])(HeaderHelp, [{
    key: "getClassNames",
    value: function getClassNames() {
      return (0, _classnames["default"])((0, _defineProperty2["default"])({
        "gd-header-help": true,
        "icon-header-help": true,
        "is-open": this.state.isOpen
      }, this.props.className, !!this.props.className));
    }
  }, {
    key: "getMenuItems",
    value: function getMenuItems() {
      var _this2 = this;

      return this.props.items.map(function (item) {
        return _react["default"].createElement("a", {
          key: item.key,
          href: item.href,
          target: item.target,
          rel: item.target === "_blank" ? "noreferrer noopener" : undefined,
          onClick: function onClick() {
            _this2.menuItemClicked(item);
          },
          className: "gd-list-item ".concat(item.className)
        }, _react["default"].createElement(_reactIntl.FormattedMessage, {
          id: item.key
        }));
      });
    }
  }, {
    key: "toggleHelpMenu",
    value: function toggleHelpMenu() {
      var isOpen = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : !this.state.isOpen;
      this.props.onHelpClicked && this.props.onHelpClicked(isOpen);
      this.setState({
        isOpen: isOpen
      });
    }
  }, {
    key: "menuItemClicked",
    value: function menuItemClicked() {
      var _this$props;

      this.toggleHelpMenu(false);

      (_this$props = this.props).onMenuItemClick.apply(_this$props, arguments);
    }
  }, {
    key: "renderHelpMenu",
    value: function renderHelpMenu() {
      var _this3 = this;

      return this.state.isOpen ? _react["default"].createElement(_Overlay["default"], {
        alignTo: ".gd-header-help",
        alignPoints: [{
          align: "br tr"
        }],
        closeOnOutsideClick: true,
        closeOnMouseDrag: true,
        closeOnParentScroll: true,
        onClose: function onClose() {
          _this3.toggleHelpMenu(false);
        }
      }, !this.props.disableDropdown ? _react["default"].createElement("div", {
        className: "gd-dialog gd-dropdown overlay gd-header-help-dropdown"
      }, _react["default"].createElement("div", {
        className: "gd-list small"
      }, this.getMenuItems())) : null) : false;
    }
  }, {
    key: "render",
    value: function render() {
      var _this4 = this;

      return _react["default"].createElement("div", {
        className: this.getClassNames(),
        onClick: function onClick() {
          return _this4.toggleHelpMenu();
        }
      }, _react["default"].createElement(_reactIntl.FormattedMessage, {
        id: "gs.header.help"
      }), this.renderHelpMenu());
    }
  }]);
  return HeaderHelp;
}(_react.PureComponent);

exports.HeaderHelp = HeaderHelp;
(0, _defineProperty2["default"])(HeaderHelp, "propTypes", {
  className: _propTypes["default"].string,
  items: _propTypes["default"].arrayOf(_HeaderMenu.itemPropType),
  onMenuItemClick: _propTypes["default"].func.isRequired,
  disableDropdown: _propTypes["default"].bool,
  onHelpClicked: _propTypes["default"].func
});
(0, _defineProperty2["default"])(HeaderHelp, "defaultProps", {
  className: "",
  items: [],
  disableDropdown: false
});

var _default = (0, _reactIntl.injectIntl)(HeaderHelp);

exports["default"] = _default;
