// @gooddata/goodstrap/lib/Header/HeaderProjectPickerItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = exports.ProjectListItem = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactIntl = require("react-intl");

var _classnames = _interopRequireDefault(require("classnames"));

var _jsUtils = require("@gooddata/js-utils");

var _intlShape = require("../core/intlShape");

// (C) 2007-2020 GoodData Corporation
var ProjectListItem = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(ProjectListItem, _Component);

  function ProjectListItem() {
    (0, _classCallCheck2["default"])(this, ProjectListItem);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(ProjectListItem).apply(this, arguments));
  }

  (0, _createClass2["default"])(ProjectListItem, [{
    key: "render",
    value: function render() {
      var _cx;

      var _this$props = this.props,
          intl = _this$props.intl,
          item = _this$props.item,
          selectedItem = _this$props.selectedItem;
      var t = intl.formatMessage;

      if (!item) {
        return _react["default"].createElement("div", {
          className: "gd-list-item gd-project-list-item gd-is-loading"
        }, _react["default"].createElement("div", {
          className: "gd-spinner small"
        }));
      }

      var isSelected = selectedItem.uri === item.uri;
      var title = item.title;
      var classes = (0, _classnames["default"])((_cx = {
        "gd-list-item": true,
        "gd-project-list-item": true
      }, (0, _defineProperty2["default"])(_cx, "s-".concat(_jsUtils.string.simplifyText(title)), true), (0, _defineProperty2["default"])(_cx, "gd-is-selected", isSelected), (0, _defineProperty2["default"])(_cx, "gd-is-demo", item.isDemo), _cx));

      var demoSticker = _react["default"].createElement("span", {
        className: "demo-sticker"
      }, t({
        id: "gs.header.projectPicker.demo"
      }));

      return _react["default"].createElement("div", {
        className: classes,
        title: title
      }, _react["default"].createElement("span", {
        className: "project-title"
      }, title), item.isDemo ? demoSticker : false);
    }
  }]);
  return ProjectListItem;
}(_react.Component);

exports.ProjectListItem = ProjectListItem;
(0, _defineProperty2["default"])(ProjectListItem, "propTypes", {
  item: _propTypes["default"].shape({
    title: _propTypes["default"].string.isRequired,
    uri: _propTypes["default"].string.isRequired
  }),
  selectedItem: _propTypes["default"].object.isRequired,
  intl: _intlShape.intlShape.isRequired
});
(0, _defineProperty2["default"])(ProjectListItem, "defaultProps", {
  item: null
});

var _default = (0, _reactIntl.injectIntl)(ProjectListItem);

exports["default"] = _default;
