// @gooddata/goodstrap/lib/Header/HeaderProjectPicker.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty2 = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty2(exports, "__esModule", {
  value: true
});

exports["default"] = exports.HeaderProjectPicker = exports.ProjectPickerButton = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/define-property"));

var _defineProperties = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/define-properties"));

var _getOwnPropertyDescriptors = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptors"));

var _getOwnPropertyDescriptor = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptor"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/get-own-property-symbols"));

var _keys = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/keys"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty3 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactIntl = require("react-intl");

var _classnames = _interopRequireDefault(require("classnames"));

var _jsUtils = require("@gooddata/js-utils");

var _Dropdown = _interopRequireWildcard(require("../Dropdown/Dropdown"));

var _NoData = _interopRequireDefault(require("../data/NoData"));

var _DataSourceWrapper = _interopRequireDefault(require("../data/DataSourceWrapper"));

var _HeaderProjectPickerItem = _interopRequireDefault(require("./HeaderProjectPickerItem"));

var _intlShape = require("../core/intlShape");

function ownKeys(object, enumerableOnly) { var keys = (0, _keys["default"])(object); if (_getOwnPropertySymbols["default"]) { var symbols = (0, _getOwnPropertySymbols["default"])(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return (0, _getOwnPropertyDescriptor["default"])(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty3["default"])(target, key, source[key]); }); } else if (_getOwnPropertyDescriptors["default"]) { (0, _defineProperties["default"])(target, (0, _getOwnPropertyDescriptors["default"])(source)); } else { ownKeys(Object(source)).forEach(function (key) { (0, _defineProperty2["default"])(target, key, (0, _getOwnPropertyDescriptor["default"])(source, key)); }); } } return target; }

var ProjectPickerButton = function ProjectPickerButton(_ref) {
  var _cx;

  var isOpen = _ref.isOpen,
      projectTitle = _ref.projectTitle;
  var classNames = (0, _classnames["default"])((_cx = {
    "gd-header-project": true
  }, (0, _defineProperty3["default"])(_cx, "s-".concat(_jsUtils.string.simplifyText(projectTitle)), true), (0, _defineProperty3["default"])(_cx, "is-expanded", isOpen), (0, _defineProperty3["default"])(_cx, "is-collapsed", !isOpen), _cx));
  return _react["default"].createElement("div", {
    className: classNames
  }, projectTitle);
};

exports.ProjectPickerButton = ProjectPickerButton;
ProjectPickerButton.propTypes = {
  isOpen: _propTypes["default"].bool,
  projectTitle: _propTypes["default"].string
};
ProjectPickerButton.defaultProps = {
  isOpen: false,
  projectTitle: ""
};

var HeaderProjectPicker = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(HeaderProjectPicker, _Component);

  function HeaderProjectPicker(props) {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2["default"])(this, HeaderProjectPicker);

    for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      args[_key - 1] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2["default"])(this, (_getPrototypeOf2 = (0, _getPrototypeOf3["default"])(HeaderProjectPicker)).call.apply(_getPrototypeOf2, [this, props].concat(args)));
    _this.state = _objectSpread({}, _this.state, {
      dataSource: _this.props.createDataSource(_this.props.profileUri)
    });
    _this.onProjectSelect = _this.onProjectSelect.bind((0, _assertThisInitialized2["default"])(_this));
    _this.renderProjectPickerFooter = _this.renderProjectPickerFooter.bind((0, _assertThisInitialized2["default"])(_this));
    return _this;
  }

  (0, _createClass2["default"])(HeaderProjectPicker, [{
    key: "onProjectSelect",
    value: function onProjectSelect(item) {
      if (item) {
        this.props.onProjectSelect(item);
      }
    }
  }, {
    key: "renderProjectPickerFooter",
    value: function renderProjectPickerFooter(closeDropdown) {
      var projectPickerFooter = this.props.projectPickerFooter;
      return projectPickerFooter ? _react["default"].createElement("div", {
        className: "gd-header-project-picker-footer"
      }, _react["default"].createElement("span", {
        onClick: closeDropdown
      }, projectPickerFooter)) : false;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          project = _this$props.project,
          intl = _this$props.intl;
      var t = intl.formatMessage;

      var NoDataComponent = _react["default"].createElement(_NoData["default"], {
        className: "s-noMatchingProjects",
        notFoundLabel: t({
          id: "gs.header.projectPicker.noMatchingProjects"
        })
      });

      return _react["default"].createElement(_Dropdown["default"], {
        closeOnParentScroll: true,
        closeOnMouseDrag: true,
        className: "gd-header-project-wrapper gd-header-measure",
        button: _react["default"].createElement(ProjectPickerButton, {
          projectTitle: project.title
        }),
        body: _react["default"].createElement(_DataSourceWrapper["default"], {
          dataSource: this.state.dataSource,
          searchPlaceholder: t({
            id: "gs.header.projectPicker.searchPlaceholder"
          })
        }, _react["default"].createElement(_Dropdown.DropdownBody, {
          className: "project-picker-dropdown",
          onSelect: this.onProjectSelect,
          width: 350,
          rowItem: _react["default"].createElement(_HeaderProjectPickerItem["default"], {
            selectedItem: project
          }),
          searchFieldSize: "normal",
          Footer: this.renderProjectPickerFooter,
          noData: NoDataComponent
        }))
      });
    }
  }]);
  return HeaderProjectPicker;
}(_react.Component);

exports.HeaderProjectPicker = HeaderProjectPicker;
(0, _defineProperty3["default"])(HeaderProjectPicker, "propTypes", {
  className: _propTypes["default"].string,
  projectPickerFooter: _propTypes["default"].node,
  createDataSource: _propTypes["default"].func.isRequired,
  intl: _intlShape.intlShape.isRequired,
  onProjectSelect: _propTypes["default"].func.isRequired,
  profileUri: _propTypes["default"].string.isRequired,
  project: _propTypes["default"].shape({
    title: _propTypes["default"].string.isRequired,
    uri: _propTypes["default"].string.isRequired,
    isDemo: _propTypes["default"].bool
  }).isRequired
});
(0, _defineProperty3["default"])(HeaderProjectPicker, "defaultProps", {
  className: ""
});

var _default = (0, _reactIntl.injectIntl)(HeaderProjectPicker);

exports["default"] = _default;
