// @gooddata/react-components/dist/internal/components/configurationPanels/UnsupportedConfigurationPanel.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
// (C) 2019 GoodData Corporation
var React = require("react");
var omit = require("lodash/omit");
var ConfigurationPanelContent_1 = require("./ConfigurationPanelContent");
var UnsupportedProperties_1 = require("../configurationControls/UnsupportedProperties");
var UnsupportedConfigurationPanel = /** @class */ (function (_super) {
    __extends(UnsupportedConfigurationPanel, _super);
    function UnsupportedConfigurationPanel() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    UnsupportedConfigurationPanel.prototype.componentDidMount = function () {
        this.props.pushData({
            properties: omit(this.props.properties, "controls"),
            references: null,
            ignoreUndoRedo: true,
        });
    };
    UnsupportedConfigurationPanel.prototype.isControlDisabled = function () {
        return true;
    };
    UnsupportedConfigurationPanel.prototype.renderConfigurationPanel = function () {
        return React.createElement(UnsupportedProperties_1.default, null);
    };
    return UnsupportedConfigurationPanel;
}(ConfigurationPanelContent_1.default));
exports.default = UnsupportedConfigurationPanel;
