// @gooddata/react-components/dist/proptypes/Afm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// (C) 2007-2018 GoodData Corporation
var PropTypes = require("prop-types");
var Filters_1 = require("./Filters");
exports.AfmPropType = PropTypes.shape({
    attributes: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.string.isRequired,
        type: PropTypes.oneOf(["date", "attribute"]).isRequired,
    })),
    filters: Filters_1.FiltersPropType,
    measures: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.string.isRequired,
        definition: PropTypes.shape({
            baseObject: PropTypes.oneOfType([
                PropTypes.shape({ id: PropTypes.string.isRequired }),
                PropTypes.shape({ lookupId: PropTypes.string.isRequired }),
            ]).isRequired,
            filters: Filters_1.FiltersPropType,
            aggregation: PropTypes.string,
            popAttribute: PropTypes.shape({
                id: PropTypes.string.isRequired,
            }),
            showInPercent: PropTypes.bool,
        }),
    })),
    nativeTotals: PropTypes.arrayOf(PropTypes.shape({
        measureIdentifier: PropTypes.string.isRequired,
        attributeIdentifiers: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
    })),
});
