// @gooddata/react-components/dist/proptypes/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// (C) 2007-2018 GoodData Corporation
var Afm_1 = require("./Afm");
exports.AfmPropType = Afm_1.AfmPropType;
var Chart_1 = require("./Chart");
exports.ChartPropTypes = Chart_1.ChartPropTypes;
var ChartConfig_1 = require("./ChartConfig");
exports.ChartConfigPropType = ChartConfig_1.ChartConfigPropType;
var DataSource_1 = require("./DataSource");
exports.DataSourcePropType = DataSource_1.DataSourcePropType;
var Events_1 = require("./Events");
exports.EventsPropTypes = Events_1.EventsPropTypes;
var Filters_1 = require("./Filters");
exports.FiltersPropType = Filters_1.FiltersPropType;
var Kpi_1 = require("./Kpi");
exports.KpiPropTypes = Kpi_1.KpiPropTypes;
var Table_1 = require("./Table");
exports.TablePropTypes = Table_1.TablePropTypes;
var Transformation_1 = require("./Transformation");
exports.TransformationPropType = Transformation_1.TransformationPropType;
var Visualization_1 = require("./Visualization");
exports.VisualizationPropType = Visualization_1.VisualizationPropType;
