// @gooddata/react-components/dist/proptypes/Transformation.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
// (C) 2007-2018 GoodData Corporation
var PropTypes = require("prop-types");
var object = {
    id: PropTypes.string.isRequired,
    title: PropTypes.string,
};
exports.TransformationPropType = PropTypes.shape({
    sorting: PropTypes.arrayOf(PropTypes.shape({
        column: PropTypes.string.isRequired,
        direction: PropTypes.string.isRequired,
    })),
    measures: PropTypes.arrayOf(PropTypes.shape(__assign({ format: PropTypes.string }, object))),
    dimensions: PropTypes.arrayOf(PropTypes.shape({
        name: PropTypes.string.isRequired,
        attributes: PropTypes.arrayOf(PropTypes.shape(object)).isRequired,
    })),
});
