// @gooddata/react-components/dist/components/afm/afmComponents.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// (C) 2007-2020 GoodData Corporation
var BarChart_1 = require("./BarChart");
exports.BarChart = BarChart_1.BarChart;
var ColumnChart_1 = require("./ColumnChart");
exports.ColumnChart = ColumnChart_1.ColumnChart;
var LineChart_1 = require("./LineChart");
exports.LineChart = LineChart_1.LineChart;
var BulletChart_1 = require("./BulletChart");
exports.BulletChart = BulletChart_1.BulletChart;
var PieChart_1 = require("./PieChart");
exports.PieChart = PieChart_1.PieChart;
var DonutChart_1 = require("./DonutChart");
exports.DonutChart = DonutChart_1.DonutChart;
var Table_1 = require("./Table");
exports.Table = Table_1.Table;
var Headline_1 = require("./Headline");
exports.Headline = Headline_1.Headline;
var Xirr_1 = require("./Xirr");
exports.Xirr = Xirr_1.Xirr;
var AreaChart_1 = require("./AreaChart");
exports.AreaChart = AreaChart_1.AreaChart;
var Treemap_1 = require("./Treemap");
exports.Treemap = Treemap_1.Treemap;
var Heatmap_1 = require("./Heatmap");
exports.Heatmap = Heatmap_1.Heatmap;
var ScatterPlot_1 = require("./ScatterPlot");
exports.ScatterPlot = ScatterPlot_1.ScatterPlot;
var DataSourceProvider_1 = require("./DataSourceProvider");
exports.dataSourceProvider = DataSourceProvider_1.dataSourceProvider;
